{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Does any transcript levels correlate with FBA-predicted metabolic fluxes?\n",
    "\n",
    "## Introduction\n",
    "\n",
    "Transcriptomics dataset collected from multiple scenarios and published in [Kannan et al 2019](https://doi.org/10.1093/insilicoplants/diz008) was provided from UoI-UC. FBA was used to predict metabolic fluxes under different scenarios and then transcript levels are compared to flux."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing transcript data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sanu/.local/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2714: DtypeWarning: Columns (0,2,3,13,14) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"/home/sanu/Downloads/gmax_leaf_CO2_normExprn_1.txt\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "229"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.columns) -17"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Seperating out different studies\n",
    "\n",
    "Data on light intensity (PPFD) during plant growth was avialable for 5 (of 8) studies. When \\[CO2\\] is not avilable, ambient levels were assumed. Where day-length is not mentioned, as assumption that daylength is 12h was assumed. In the case of studies involving biotic and abiotic treatments (other than high \\[CO2\\] treatments), only the WT/control data was used in this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "amb_list = list(df.columns[0:17])\n",
    "ele_list = list(df.columns[0:17])\n",
    "for x in df.columns:\n",
    "    if \"Amb\" in x:\n",
    "        amb_list.append(x)\n",
    "    elif \"Ele\" in x:\n",
    "        ele_list.append(x)\n",
    "    else:\n",
    "        continue\n",
    "\n",
    "budRemove_list = list(df.columns[0:17])+[\"GSM569808.CEL\",\"GSM569809.CEL\",\"GSM569810.CEL\",\n",
    "                                         \"GSM569811.CEL\",\"GSM569812.CEL\",\"GSM569813.CEL\",\n",
    "                                         \"GSM569814.CEL\"]\n",
    "heatShock_list = list(df.columns[0:17])+[\"GSM643115.CEL\",\"GSM643116.CEL\",\"GSM643117.CEL\",\"GSM643118.CEL\"]\n",
    "mycInfect_list = list(df.columns[0:17])+[\"GSM1088939_Schaar_19_wt_c1_S.CEL\",\n",
    "                                         \"GSM1088940_Schaar_20_wt_c2_S.CEL\",\n",
    "                                         \"GSM1088954_Schaar_25_nts_c1_S.CEL\",\n",
    "                                         \"GSM1088955_Schaar_26_nts_c2_S.CEL\",\n",
    "                                         \"GSM1088956_Schaar_27_nts_c3_S.CEL\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_amb = df[amb_list]\n",
    "df_ele = df[ele_list]\n",
    "df_budRemove = df[budRemove_list]\n",
    "df_heatShock = df[heatShock_list]\n",
    "df_mycInfect = df[mycInfect_list]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set up leaf model and add gene-protein-reaction (GPR) associations from PlantCyc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra import io\n",
    "from cobra.core import Metabolite\n",
    "from cobra import flux_analysis\n",
    "from IPython import display\n",
    "import logging\n",
    "logging.basicConfig()\n",
    "logger = logging.getLogger('logger')\n",
    "\n",
    "#import sbml file\n",
    "fname = \"/home/sanu/Documents/Scripts/git/plantcoremetabolism-model/PlantCoreMetabolism_v1_3_0.xml\"\n",
    "model = io.sbml.read_sbml_model(fname)\n",
    "display.clear_output()\n",
    "\n",
    "from sweetlovegroup.transform import fixModelCompatibilityIssueCobra015\n",
    "model = fixModelCompatibilityIssueCobra015(model,fname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "892"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "861"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pythoncyc\n",
    "from Functions import *\n",
    "\n",
    "soy = pythoncyc.select_organism(\"soy\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------\n",
      "This list of metabolic reactions are ignored\n",
      "['RXN_9650_p', '2_KETO_ADIPATE_DEHYDROG_RXN_m', 'Phytol_biosynthesis_p', 'CYSTEINE_AMINOTRANSFERASE_RXN_m', 'GLYCINE_TRNA_LIGASE_RXN_c', 'RXN66_1_c', 'RXN_9648_p', 'RXN-9651', 'Plastidial_ATP_Synthase_p', 'GGPP_biosynthesis_p', 'RXN_9653_p', 'lycopene_biosynthesis_p', 'RXN_2141_p', 'SUCCINYL_COA_HYDROLASE_RXN_m', 'PROTON_ATPase_c', 'MDA_Fd_Ascorbate_p', 'MercaptoPyruvateSulfurtransferase_m', 'Phytol_degradation_p', 'RXN_9652_p', 'A_B_oxidation_x', 'unlProtHYPO_c', 'Mitochondrial_ATP_Synthase_m', 'IPP_biosynthesis_c', 'Mehler_Reaction_p', 'Beta_Oxidation_x', 'HMBPP_synthesis_p', 'OROTATE_REDUCTASE_NADH_RXN_p', 'Ferredoxin_Plastoquinone_Reductase_p', 'RXN_9651_p', 'NADPH_Dehydrogenase_p', 'Plastoquinol_Oxidase_p', 'SUCCINATE_COA_LIGASE_GDP_FORMING_RXN_m', 'RXN_1781_v', 'PREPHENATE_DEHYDROGENASE_NADP_RXN_p', 'PREPHENATEDEHYDROG_RXN_p', 'MALEYLACETOACETATE_ISOMERASE_RXN_c', 'RXN_9654_p', 'LCYSDESULF_RXN_c', 'RXN_9958_NAD_m', 'HEXOKINASE_RXN_MANNOSE_c', 'PYRUVDEH_RXN_p', 'PYRUVDEH_RXN_m']\n",
      "--------------\n",
      "RXN_1826_p\n",
      "RXN_12486_c\n",
      "arachidate_biosynthesis_p\n",
      "docosanoate_biosynthesis_p\n",
      "gadoleic_acid_biosynthesis_p\n",
      "Arachidoyl_ACP_desaturase_p\n",
      "LPG_biosynthesis_c\n",
      "RXN_6884_m\n",
      "PHE_NH4_Lyase_c\n",
      "RXN_8871_c\n",
      "RXN490_3650_p\n",
      "Octadecadienoyl_ACP_desaturase_p\n",
      "RXN_8873_c\n",
      "ACETOVANILLONE_HYDROXYLASE_c\n",
      "6PGLUCONDEHYDROG_RXN_c\n",
      "Octadecadienoyl_ACP_hydrolase_p\n",
      "Octadecatrienoyl_ACP_hydrolase_p\n",
      "palmitoyl_ACP_desaturase_p\n",
      "RXN1G_368_p\n",
      "palmitoleoyl_ACP_desaturase_p\n",
      "6PGLUCONDEHYDROG_RXN_p\n",
      "hexadecadienoate_biosynthesis_p\n",
      "hexadecadienoate_ACP_desaturase_p\n",
      "RXN1G_445_p\n",
      "hexadecatrienoate_biosynthesis_p\n",
      "RXN_8872_c\n",
      "SYRINGICACID_BIOSYNTHESIS_c\n",
      "ACETOVANILLONE_SYNTHASE_c\n",
      "Oleoyl_ACP_desaturase_p\n",
      "VANILLICACID_BIOSYNTHESIS_c\n",
      "ACETOSYRINGONE_BIOSYNTHESIS_c\n",
      "hexadecenoate_biosynthesis_p\n",
      "SYRINGALDEHYDE_BIOSYNTHESIS_c\n"
     ]
    }
   ],
   "source": [
    "model = addGPR2Models(model,soy)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modelling fluxes under different scenarios\n",
    "\n",
    "Create leaf model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "########################################################\n",
    "#This function was used to set up a C3 leaf diel model #\n",
    "########################################################\n",
    "def setupC3DielModel(core_model,transferMets,starch_sucrose_ratio=90.0/100):\n",
    "    from cobra.core import Metabolite, Reaction\n",
    "    import re\n",
    "\n",
    "    #create two copies of model elements for day and night\n",
    "    cobra_model2 = core_model.copy()\n",
    "    for met in cobra_model2.metabolites:\n",
    "        met.id = met.id+\"1\"\n",
    "        met.compartment = met.compartment+\"1\"\n",
    "    for rxn in cobra_model2.reactions:\n",
    "        rxn.id = rxn.id+\"1\"\n",
    "\n",
    "    cobra_model3 = core_model.copy()\n",
    "    for met in cobra_model3.metabolites:\n",
    "        met.id = met.id+\"2\"\n",
    "        met.compartment = met.compartment+\"2\"\n",
    "    for rxn in cobra_model3.reactions:\n",
    "        rxn.id = rxn.id+\"2\"\n",
    "\n",
    "    #merge the day and night model\n",
    "    cobra_model = cobra_model2+cobra_model3\n",
    "    for met in cobra_model3.metabolites:\n",
    "        if not cobra_model.metabolites.__contains__(met.id):\n",
    "            cobra_model.add_metabolites(met.copy())\n",
    "\n",
    "    met1 = Metabolite(\"X_Phloem_contribution_t1\",name=\"Phloem output during the day\")\n",
    "    cobra_model.reactions.get_by_id(\"Phloem_output_tx1\").add_metabolites({met1:1})\n",
    "    met2 = Metabolite(\"X_Phloem_contribution_t2\",name=\"Phloem output during at night\")\n",
    "    cobra_model.reactions.get_by_id(\"Phloem_output_tx2\").add_metabolites({met2:1})\n",
    "\n",
    "    rxn = Reaction(\"diel_biomass\")\n",
    "    rxn.add_metabolites({met1:-3,met2:-1})\n",
    "    rxn.lower_bound = 0\n",
    "    rxn.upper_bound = 1000\n",
    "    cobra_model.add_reaction(rxn)\n",
    "\n",
    "    #Adding reactions to allow for day-night metabolite accumulations\n",
    "    tmfile = open(transferMets,\"r\")\n",
    "    tmset=set()\n",
    "    for line in tmfile:\n",
    "        tmset.add(line.replace(\"\\n\",\"\"))\n",
    "\n",
    "    for met in tmset:\n",
    "        if met == \"AMMONIUM_v\" or met==\"FRUCTAN_v\":\n",
    "            continue\n",
    "        tempRxn = Reaction(met+\"_dielTransfer\")\n",
    "        tempRxn.add_metabolites({cobra_model.metabolites.get_by_id(met+\"1\"):-1,cobra_model.metabolites.get_by_id(met+\"2\"):1})\n",
    "        tempRxn.lower_bound=-1000\n",
    "        if not ((met == \"STARCH_p\") or (met == \"SUCROSE_v\") or (met == \"MAL_v\") or (met == \"aMAL_v\") or (met == \"NITRATE_v\") or (met == \"CIT_v\") or (met == \"aCIT_v\") or (met == \"PROTON_v\")):\n",
    "            tempRxn.lower_bound=0\n",
    "        tempRxn.upper_bound=1000\n",
    "        cobra_model.add_reaction(tempRxn)\n",
    "\n",
    "    fractionMets=dict()\n",
    "    for rxn in cobra_model.reactions:\n",
    "        for met in rxn.metabolites.keys():\n",
    "            prefix=\"\"\n",
    "            a=re.search(\"^a{1,3}\",met.id)\n",
    "            anion=\"\"\n",
    "            if a:\n",
    "                anion=a.group(0)\n",
    "                prefix=anion\n",
    "            b=re.search(\"^b{1,3}\",met.id)\n",
    "            basic=\"\"\n",
    "            if b:\n",
    "                basic=b.group(0)\n",
    "                prefix=basic\n",
    "            if ((not prefix == \"\") and met.compartment == \"v1\"):\n",
    "                fractionMets[met]=prefix\n",
    "\n",
    "    temp=cobra_model.copy()\n",
    "    for met in fractionMets.keys():\n",
    "        for rxn in met.reactions:\n",
    "            if rxn.id.__contains__(\"_dielTransfer\"):\n",
    "                continue\n",
    "            else:\n",
    "                mainMet = met.id[len(fractionMets[met]):]\n",
    "                coeff1 = temp.reactions.get_by_id(rxn.id).metabolites.get(temp.metabolites.get_by_id(mainMet))\n",
    "                coeff2 = temp.reactions.get_by_id(rxn.id).metabolites.get(temp.metabolites.get_by_id(met.id))\n",
    "                if not coeff1:\n",
    "                    coeff1=0\n",
    "                if not coeff2:\n",
    "                    coeff2=0\n",
    "                total = coeff1 + coeff2\n",
    "                coeff1 = float(coeff1)/total\n",
    "                coeff2 = float(coeff2)/total\n",
    "                if cobra_model.reactions.has_id(met.id[0:len(met.id)-1]+\"_dielTransfer\"):\n",
    "                    ub = temp.reactions.get_by_id(met.id[0:len(met.id)-1]+\"_dielTransfer\").upper_bound\n",
    "                    lb = temp.reactions.get_by_id(met.id[0:len(met.id)-1]+\"_dielTransfer\").lower_bound\n",
    "                    temp.reactions.get_by_id(met.id[0:len(met.id)-1]+\"_dielTransfer\").remove_from_model()\n",
    "                    temp.reactions.get_by_id(mainMet[0:len(mainMet)-1]+\"_dielTransfer\").remove_from_model()\n",
    "                    Reac = Reaction(mainMet[0:len(mainMet)-1]+\"_dielTransfer\",name=mainMet+\"_dielTransfer\")\n",
    "                    Reac.add_metabolites({temp.metabolites.get_by_id(met.id[0:len(met.id)-1]+\"1\"):-coeff2,temp.metabolites.get_by_id(met.id[0:len(met.id)-1]+\"2\"):coeff2,temp.metabolites.get_by_id(mainMet[0:len(mainMet)-1]+\"1\"):-coeff1,temp.metabolites.get_by_id(mainMet[0:len(mainMet)-1]+\"2\"):coeff1})\n",
    "                    Reac.lower_bound=lb\n",
    "                    Reac.upper_bound=ub\n",
    "                    temp.add_reaction(Reac)\n",
    "                    print Reac.reaction\n",
    "                break\n",
    "    ####ADD CONSTRAINTS TO MODEL####\n",
    "    cobra_model = temp.copy()\n",
    "\n",
    "    #objective function\n",
    "    cobra_model.reactions.get_by_id(\"diel_biomass\").objective_coefficient=1\n",
    "    #Leaves - light\n",
    "    cobra_model.reactions.get_by_id(\"Sucrose_tx1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"Sucrose_tx1\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"GLC_tx1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"GLC_tx1\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"CO2_tx1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NH4_tx1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NH4_tx1\").upper_bound=0\n",
    "    #Leaves - dark\n",
    "    cobra_model.reactions.get_by_id(\"Sucrose_tx2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"Sucrose_tx2\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"GLC_tx2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"GLC_tx2\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"Photon_tx2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"Photon_tx2\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NH4_tx2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NH4_tx2\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"CO2_tx2\").upper_bound=0\n",
    "\n",
    "    #Set pG6P transporter to 0\n",
    "    cobra_model.reactions.get_by_id(\"G6P_Pi_pc1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"G6P_Pi_pc1\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"G6P_Pi_pc2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"G6P_Pi_pc2\").upper_bound=0\n",
    "\n",
    "    #Turn off PTOX\n",
    "    cobra_model.reactions.get_by_id(\"Plastoquinol_Oxidase_p1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"Plastoquinol_Oxidase_p1\").upper_bound=0\n",
    "\n",
    "    #nitrate uptake constrain\n",
    "    Nitrate_balance = Metabolite(\"Nitrate_bal_c\", name = \"Weights to balance nitrate uptake\", compartment = \"c1\")\n",
    "    cobra_model.reactions.get_by_id(\"Nitrate_ec1\").add_metabolites({Nitrate_balance:-2})\n",
    "    cobra_model.reactions.get_by_id(\"Nitrate_ec2\").add_metabolites({Nitrate_balance:3})\n",
    "\n",
    "    #Rubisco balance\n",
    "    Rubisco_balance = Metabolite(\"rubisco_bal_p1\", name = \"Weights to balance RuBP carboxygenase oxygenase balance\", compartment = \"p1\")\n",
    "    cobra_model.reactions.get_by_id(\"RXN_961_p1\").add_metabolites({Rubisco_balance:3})\n",
    "    cobra_model.reactions.get_by_id(\"RIBULOSE_BISPHOSPHATE_CARBOXYLASE_RXN_p1\").add_metabolites({Rubisco_balance:-1})\n",
    "\n",
    "    #generic ATPase and NADPH oxidase\n",
    "    Maintenance_constraint = Metabolite(\"ATPase_NADPHoxidase_constraint_c1\",name =  \"ATPase_NADPHoxidase_constraint_c1\", compartment = \"c1\")\n",
    "    Maintenance_constraint2 = Metabolite(\"ATPase_NADPHoxidase_constraint_c2\",name =  \"ATPase_NADPHoxidase_constraint_c2\", compartment = \"c2\")\n",
    "    Maintenance_constraint3 = Metabolite(\"Light_dark_maintainence_constraint\",name =  \"Light_dark_maintainence_constraint\", compartment = \"c1\")\n",
    "    cobra_model.reactions.get_by_id(\"ATPase_tx1\").add_metabolites({Maintenance_constraint:1,Maintenance_constraint3:1})\n",
    "    cobra_model.reactions.get_by_id(\"ATPase_tx2\").add_metabolites({Maintenance_constraint2:1,Maintenance_constraint3:-1})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxc_tx1\").add_metabolites({Maintenance_constraint:-3})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxc_tx2\").add_metabolites({Maintenance_constraint2:-3})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxm_tx1\").add_metabolites({Maintenance_constraint:-3})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxm_tx2\").add_metabolites({Maintenance_constraint2:-3})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxp_tx1\").add_metabolites({Maintenance_constraint:-3})\n",
    "    cobra_model.reactions.get_by_id(\"NADPHoxp_tx2\").add_metabolites({Maintenance_constraint2:-3})\n",
    "\n",
    "    ##constrain sucrose and starch storage\n",
    "    Sucorse_starch_balance = Metabolite(\"sucrose_starch_bal_c\", name = \"Weights to balance sucrose-starch uptake\", compartment = \"c1\")\n",
    "    cobra_model.reactions.get_by_id(\"SUCROSE_v_dielTransfer\").add_metabolites({Sucorse_starch_balance:-1*starch_sucrose_ratio})\n",
    "    cobra_model.reactions.get_by_id(\"STARCH_p_dielTransfer\").add_metabolites({Sucorse_starch_balance:1})\n",
    "\n",
    "    #Plastid enolase was not detected in Arabidopsis mesophyll tissue\n",
    "    cobra_model.reactions.get_by_id(\"2PGADEHYDRAT_RXN_p1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"2PGADEHYDRAT_RXN_p1\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"2PGADEHYDRAT_RXN_p2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"2PGADEHYDRAT_RXN_p2\").upper_bound=0\n",
    "\n",
    "    #Setting chloroplastic NADPH dehydrogenase to 0  ((Yamamoto et al., 2011)\n",
    "    cobra_model.reactions.get_by_id(\"NADPH_Dehydrogenase_p1\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NADPH_Dehydrogenase_p1\").upper_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NADPH_Dehydrogenase_p2\").lower_bound=0\n",
    "    cobra_model.reactions.get_by_id(\"NADPH_Dehydrogenase_p2\").upper_bound=0\n",
    "\n",
    "    #ATP_ADP_Pi constrained to 0 because while there is evidence for its existance, it does not carry flux during the day\n",
    "    cobra_model.reactions.get_by_id(\"ATP_ADP_Pi_pc1\").lower_bound = 0\n",
    "    cobra_model.reactions.get_by_id(\"ATP_ADP_Pi_pc1\").upper_bound = 0\n",
    "\n",
    "    return cobra_model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.7 MAL_v1 + 0.3 aMAL_v1 <=> 0.7 MAL_v2 + 0.3 aMAL_v2\n",
      "0.5 CIT_v1 + 0.5 aCIT_v1 <=> 0.5 CIT_v2 + 0.5 aCIT_v2\n",
      "bHIS_v1 --> bHIS_v2\n"
     ]
    }
   ],
   "source": [
    "leaf_model = model.copy()\n",
    "\n",
    "from sweetlovegroup import transform\n",
    "leaf_model = setupC3DielModel(leaf_model,\"MetabolitesToTransfer.txt\",starch_sucrose_ratio=5)\n",
    "\n",
    "for rxn in leaf_model.reactions:\n",
    "    if rxn.lower_bound==-1000:\n",
    "        rxn.lower_bound=-2000\n",
    "    if rxn.upper_bound==1000:\n",
    "        rxn.upper_bound=2000\n",
    "        \n",
    "leaf_model.reactions.diel_biomass.objective_coefficient=1\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scenario 1\n",
    "\n",
    "PPFD = 1500 ([Kannan et al 2019](https://doi.org/10.1093/insilicoplants/diz008))  \n",
    "\\[CO2\\] = 380  ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))  \n",
    "net CO2 uptake rate = 29.85 ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))  \n",
    "daylength = 14h ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))  \n",
    "\n",
    "\n",
    "|sampling time|day length|\n",
    "|-|-|\n",
    "|11/07/2005|14h50m|\n",
    "|01/08/2005|14h17m|\n",
    "|02/08/2005|14h15m|\n",
    "|23/08/2005|13h28m|\n",
    "|24/08/2005|13h25m|\n",
    "|07/09/2005|12h50m|\n",
    "|18/07/2006|14h42m|\n",
    "|02/08/2006|14h16m|\n",
    "|25/08/2006|13h24m|\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75.6\n",
      "0.0504\n",
      "1.50444\n",
      "Phloem output = \n",
      "0.0281640962762\n"
     ]
    }
   ],
   "source": [
    "PPFD = 1500\n",
    "daylength = 14\n",
    "ATPase = (0.0049*PPFD) + 2.7851\n",
    "scale = float(60*60*daylength)/1000000\n",
    "\n",
    "from sweetlovegroup import analysis\n",
    "temp = leaf_model.copy()\n",
    "from Functions import adjustObjectiveBasedOnDaylength\n",
    "temp = adjustObjectiveBasedOnDaylength(temp,daylength)\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "out_400 = analysis.estimateOutputFromNetCO2(temp,29.85*scale)\n",
    "\n",
    "leaf_model_400 = leaf_model.copy()\n",
    "leaf_model_400 = adjustObjectiveBasedOnDaylength(leaf_model_400,daylength)\n",
    "leaf_model_400.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "leaf_model_400.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "leaf_model_400.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "leaf_model_400.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "leaf_model_400.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "leaf_model_400.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "leaf_model_400.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "\n",
    "leaf_model_400.reactions.diel_biomass.lower_bound = out_400\n",
    "leaf_model_400.reactions.diel_biomass.upper_bound = out_400\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model_400)\n",
    "print(\"Phloem output = \")\n",
    "print(leaf_model_400.reactions.diel_biomass.flux)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "totalTranscript_400 = dict()\n",
    "for rxn in leaf_model.reactions:\n",
    "    #print rxn.id\n",
    "    count = 0\n",
    "    x = np.zeros((1,len(df_amb.columns[17:])))\n",
    "    for gene in rxn.genes:\n",
    "        if gene.id in list(df_amb[\"Glyma2.0\"]):\n",
    "            #print gene\n",
    "            if count==0:\n",
    "                x = df_amb[df_amb[\"Glyma2.0\"]==gene.id][df_amb.columns[17:]].values\n",
    "                if len(x)>1:\n",
    "                    for i in range(0,len(x)):\n",
    "                        if i == 0:\n",
    "                            x2 = x[i]\n",
    "                        else:\n",
    "                            x2 = x2 + x[i]\n",
    "                    x = np.array([x2])\n",
    "                count = count+1\n",
    "            else:\n",
    "                #print x\n",
    "                y = df_amb[df_amb[\"Glyma2.0\"]==gene.id][df_amb.columns[17:]].values\n",
    "                if len(y)>1:\n",
    "                    for i in range(0,len(y)):\n",
    "                        if i==0:\n",
    "                            y2 = y[i]\n",
    "                        else:\n",
    "                            y2 = y2 + y[i]\n",
    "                    y = np.array([y2])\n",
    "                x = x + y\n",
    "                count = count+1\n",
    "                #print x\n",
    "    #print count\n",
    "    totalTranscript_400[rxn.id]=x[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scenario 2\n",
    "\n",
    "PPFD = 1500 ([Kannan et al 2019](https://doi.org/10.1093/insilicoplants/diz008))    \n",
    "\\[CO2\\] = 550 ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))  \n",
    "net CO2 uptake rate = 36.36 ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))  \n",
    "daylength = 14h ([Leakey et al 2009](https://doi.org/10.1073/pnas.0810955106))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "75.6\n",
      "0.0504\n",
      "1.832544\n",
      "Phloem output = \n",
      "0.0339353990711\n"
     ]
    }
   ],
   "source": [
    "PPFD = 1500\n",
    "daylength = 14\n",
    "ATPase = (0.0049*PPFD) + 2.7851\n",
    "scale = float(60*60*daylength)/1000000\n",
    "\n",
    "from sweetlovegroup import analysis\n",
    "temp = leaf_model.copy()\n",
    "from Functions import adjustObjectiveBasedOnDaylength\n",
    "temp = adjustObjectiveBasedOnDaylength(temp,daylength)\n",
    "temp.reactions.get_by_id(\"RXN_961_p1\").add_metabolites({temp.metabolites.get_by_id(\"rubisco_bal_p1\"):1.35})\n",
    "\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "out_550 = analysis.estimateOutputFromNetCO2(temp,36.36*scale)\n",
    "\n",
    "leaf_model_550 = leaf_model.copy()\n",
    "leaf_model_550 = adjustObjectiveBasedOnDaylength(leaf_model_550,daylength)\n",
    "leaf_model_550.reactions.get_by_id(\"RXN_961_p1\").add_metabolites({leaf_model_550.metabolites.get_by_id(\"rubisco_bal_p1\"):1.35})\n",
    "leaf_model_550.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "leaf_model_550.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "leaf_model_550.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "leaf_model_550.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "leaf_model_550.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "leaf_model_550.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "leaf_model_550.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "leaf_model_550.reactions.diel_biomass.lower_bound = out_550\n",
    "leaf_model_550.reactions.diel_biomass.upper_bound = out_550\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model_550)\n",
    "print(\"Phloem output = \")\n",
    "print(leaf_model_550.reactions.diel_biomass.flux)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "totalTranscript_550 = dict()\n",
    "temp = df_ele\n",
    "\n",
    "for rxn in leaf_model.reactions:\n",
    "    #print rxn.id\n",
    "    count = 0\n",
    "    x = np.zeros((1,len(temp.columns[17:])))\n",
    "    for gene in rxn.genes:\n",
    "        if gene.id in list(temp[\"Glyma2.0\"]):\n",
    "            #print gene\n",
    "            if count==0:\n",
    "                x = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(x)>1:\n",
    "                    for i in range(0,len(x)):\n",
    "                        if i == 0:\n",
    "                            x2 = x[i]\n",
    "                        else:\n",
    "                            x2 = x2 + x[i]\n",
    "                    x = np.array([x2])\n",
    "                count = count+1\n",
    "            else:\n",
    "                #print x\n",
    "                y = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(y)>1:\n",
    "                    for i in range(0,len(y)):\n",
    "                        if i==0:\n",
    "                            y2 = y[i]\n",
    "                        else:\n",
    "                            y2 = y2 + y[i]\n",
    "                    y = np.array([y2])\n",
    "                x = x + y\n",
    "                count = count+1\n",
    "                #print x\n",
    "    #print count\n",
    "    totalTranscript_550[rxn.id]=x[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scenario 3\n",
    "\n",
    "PPFD = 400 ([Turner et al 2012](https://doi.org/10.1007/s00425-011-1551-4))  \n",
    "\\[CO2\\] = 370  (ambient [CO2] assumed)  \n",
    "net CO2 uptake rate = 23.86 (Predicted according to [Kannan et al 2019](https://doi.org/10.1093/insilicoplants/diz008))  \n",
    "daylength = 16h ([Turner et al 2012](https://doi.org/10.1007/s00425-011-1551-4))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23.04\n",
      "0.0576\n",
      "1.374336\n",
      "Phloem output = \n",
      "0.0190870137859\n"
     ]
    }
   ],
   "source": [
    "PPFD = 400\n",
    "daylength = 16\n",
    "ATPase = (0.0049*PPFD) + 2.7851\n",
    "scale = float(60*60*daylength)/1000000\n",
    "\n",
    "from sweetlovegroup import analysis\n",
    "temp = leaf_model.copy()\n",
    "from Functions import adjustObjectiveBasedOnDaylength\n",
    "temp = adjustObjectiveBasedOnDaylength(temp,daylength)\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "out_400 = analysis.estimateOutputFromNetCO2(temp,23.86*scale)\n",
    "\n",
    "leaf_model_budRemove = leaf_model.copy()\n",
    "leaf_model_budRemove = adjustObjectiveBasedOnDaylength(leaf_model_budRemove,daylength)\n",
    "leaf_model_budRemove.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "leaf_model_budRemove.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "leaf_model_budRemove.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "leaf_model_budRemove.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "leaf_model_budRemove.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "leaf_model_budRemove.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "leaf_model_budRemove.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "leaf_model_budRemove.reactions.diel_biomass.lower_bound = out_400\n",
    "leaf_model_budRemove.reactions.diel_biomass.upper_bound = out_400\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model_budRemove)\n",
    "print(\"Phloem output = \")\n",
    "print(leaf_model_budRemove.reactions.diel_biomass.flux)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "totalTranscript_budRemove = dict()\n",
    "temp = df_budRemove\n",
    "\n",
    "for rxn in leaf_model.reactions:\n",
    "    #print rxn.id\n",
    "    count = 0\n",
    "    x = np.zeros((1,len(temp.columns[17:])))\n",
    "    for gene in rxn.genes:\n",
    "        if gene.id in list(temp[\"Glyma2.0\"]):\n",
    "            #print gene\n",
    "            if count==0:\n",
    "                x = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(x)>1:\n",
    "                    for i in range(0,len(x)):\n",
    "                        if i == 0:\n",
    "                            x2 = x[i]\n",
    "                        else:\n",
    "                            x2 = x2 + x[i]\n",
    "                    x = np.array([x2])\n",
    "                count = count+1\n",
    "            else:\n",
    "                #print x\n",
    "                y = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(y)>1:\n",
    "                    for i in range(0,len(y)):\n",
    "                        if i==0:\n",
    "                            y2 = y[i]\n",
    "                        else:\n",
    "                            y2 = y2 + y[i]\n",
    "                    y = np.array([y2])\n",
    "                x = x + y\n",
    "                count = count+1\n",
    "                #print x\n",
    "    #print count\n",
    "    totalTranscript_budRemove[rxn.id]=x[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scenario 4\n",
    "\n",
    "PPFD = 350 ([Weston et al 2011](https://doi.org/10.1111/j.1365-3040.2011.02347.x))  \n",
    "\\[CO2\\] = 380 ([Weston et al 2011](https://doi.org/10.1111/j.1365-3040.2011.02347.x))  \n",
    "net CO2 uptake rate = 13.98 (Average of 14.55 and 13.40 \\[CO2\\]=370) ([Haile nd Higley 2003](https://doi.org/10.1603/0046-225x-32.3.433))  \n",
    "daylength = 10h ([Weston et al 2011](https://doi.org/10.1111/j.1365-3040.2011.02347.x))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12.6\n",
      "0.036\n",
      "0.50328\n",
      "Phloem output = \n",
      "0.0153346274356\n"
     ]
    }
   ],
   "source": [
    "PPFD = 350\n",
    "daylength = 10\n",
    "ATPase = (0.0049*PPFD) + 2.7851\n",
    "scale = float(60*60*daylength)/1000000\n",
    "\n",
    "from sweetlovegroup import analysis\n",
    "temp = leaf_model.copy()\n",
    "from Functions import adjustObjectiveBasedOnDaylength\n",
    "temp = adjustObjectiveBasedOnDaylength(temp,daylength)\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "out_370 = analysis.estimateOutputFromNetCO2(temp,13.98*scale)\n",
    "\n",
    "leaf_model_heatShock = leaf_model.copy()\n",
    "leaf_model_heatShock = adjustObjectiveBasedOnDaylength(leaf_model_heatShock,daylength)\n",
    "leaf_model_heatShock.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "leaf_model_heatShock.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "leaf_model_heatShock.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "leaf_model_heatShock.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "leaf_model_heatShock.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "leaf_model_heatShock.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "leaf_model_heatShock.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "leaf_model_heatShock.reactions.diel_biomass.lower_bound = out_370\n",
    "leaf_model_heatShock.reactions.diel_biomass.upper_bound = out_370\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model_heatShock)\n",
    "print(\"Phloem output = \")\n",
    "print(leaf_model_heatShock.reactions.diel_biomass.flux)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "totalTranscript_heatShock = dict()\n",
    "temp = df_heatShock\n",
    "\n",
    "for rxn in leaf_model.reactions:\n",
    "    #print rxn.id\n",
    "    count = 0\n",
    "    x = np.zeros((1,len(temp.columns[17:])))\n",
    "    for gene in rxn.genes:\n",
    "        if gene.id in list(temp[\"Glyma2.0\"]):\n",
    "            #print gene\n",
    "            if count==0:\n",
    "                x = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(x)>1:\n",
    "                    for i in range(0,len(x)):\n",
    "                        if i == 0:\n",
    "                            x2 = x[i]\n",
    "                        else:\n",
    "                            x2 = x2 + x[i]\n",
    "                    x = np.array([x2])\n",
    "                count = count+1\n",
    "            else:\n",
    "                #print x\n",
    "                y = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(y)>1:\n",
    "                    for i in range(0,len(y)):\n",
    "                        if i==0:\n",
    "                            y2 = y[i]\n",
    "                        else:\n",
    "                            y2 = y2 + y[i]\n",
    "                    y = np.array([y2])\n",
    "                x = x + y\n",
    "                count = count+1\n",
    "                #print x\n",
    "    #print count\n",
    "    totalTranscript_heatShock[rxn.id]=x[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scenario 5\n",
    "\n",
    "PPFD = 200  \n",
    "\\[CO2\\] = ? , assuming 390 based on publication date and CO2 data from the year (https://www.co2.earth/monthly-co2)  \n",
    "net CO2 uptake rate = 13.73 (Predicted according to [Kannan et al 2019](https://doi.org/10.1093/insilicoplants/diz008) which models [CO2]=370)  \n",
    "daylength = 16h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11.52\n",
      "0.0576\n",
      "0.790848\n",
      "Phloem output = \n",
      "0.0108700826235\n"
     ]
    }
   ],
   "source": [
    "PPFD = 200\n",
    "daylength = 16\n",
    "ATPase = (0.0049*PPFD) + 2.7851\n",
    "scale = float(60*60*daylength)/1000000\n",
    "\n",
    "from sweetlovegroup import analysis\n",
    "temp = leaf_model.copy()\n",
    "from Functions import adjustObjectiveBasedOnDaylength\n",
    "temp = adjustObjectiveBasedOnDaylength(temp,daylength)\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "temp.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "temp.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "temp.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "out_370 = analysis.estimateOutputFromNetCO2(temp,13.73*scale)\n",
    "\n",
    "leaf_model_mycInfect = leaf_model.copy()\n",
    "leaf_model_mycInfect = adjustObjectiveBasedOnDaylength(leaf_model_mycInfect,daylength)\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"Photon_tx1\").lower_bound = 0\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"Photon_tx1\").upper_bound = PPFD*scale\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"ATPase_tx1\").lower_bound = 1*scale\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"ATPase_tx1\").upper_bound = 1*scale\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_p1\").upper_bound = 0.034*scale\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"OAA_MAL_pc1\").upper_bound = 0.75*scale\n",
    "leaf_model_mycInfect.reactions.get_by_id(\"OAA_MAL_pc1\").lower_bound = -0.75*scale\n",
    "leaf_model_mycInfect.reactions.diel_biomass.lower_bound = out_370\n",
    "leaf_model_mycInfect.reactions.diel_biomass.upper_bound = out_370\n",
    "sol = flux_analysis.parsimonious.pfba(leaf_model_mycInfect)\n",
    "print(\"Phloem output = \")\n",
    "print(leaf_model_mycInfect.reactions.diel_biomass.flux)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0432"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "0.75*scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.7500000000000001"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leaf_model_mycInfect.reactions.get_by_id(\"OAA_MAL_pc1\").flux/scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "totalTranscript_mycInfect = dict()\n",
    "temp = df_mycInfect\n",
    "\n",
    "for rxn in leaf_model.reactions:\n",
    "    #print rxn.id\n",
    "    count = 0\n",
    "    x = np.zeros((1,len(temp.columns[17:])))\n",
    "    for gene in rxn.genes:\n",
    "        if gene.id in list(temp[\"Glyma2.0\"]):\n",
    "            #print gene\n",
    "            if count==0:\n",
    "                x = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(x)>1:\n",
    "                    for i in range(0,len(x)):\n",
    "                        if i == 0:\n",
    "                            x2 = x[i]\n",
    "                        else:\n",
    "                            x2 = x2 + x[i]\n",
    "                    x = np.array([x2])\n",
    "                count = count+1\n",
    "            else:\n",
    "                #print x\n",
    "                y = temp[temp[\"Glyma2.0\"]==gene.id][temp.columns[17:]].values\n",
    "                if len(y)>1:\n",
    "                    for i in range(0,len(y)):\n",
    "                        if i==0:\n",
    "                            y2 = y[i]\n",
    "                        else:\n",
    "                            y2 = y2 + y[i]\n",
    "                    y = np.array([y2])\n",
    "                x = x + y\n",
    "                count = count+1\n",
    "                #print x\n",
    "    #print count\n",
    "    totalTranscript_mycInfect[rxn.id]=x[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "modelDict = dict()\n",
    "modelDict[1] = leaf_model_400\n",
    "modelDict[2] = leaf_model_550\n",
    "modelDict[3] = leaf_model_budRemove\n",
    "modelDict[4] = leaf_model_heatShock\n",
    "modelDict[5] = leaf_model_mycInfect\n",
    "\n",
    "totalTranscriptDict = dict()\n",
    "totalTranscriptDict[1] = totalTranscript_400\n",
    "totalTranscriptDict[2] = totalTranscript_550\n",
    "totalTranscriptDict[3] = totalTranscript_budRemove\n",
    "totalTranscriptDict[4] = totalTranscript_heatShock\n",
    "totalTranscriptDict[5] = totalTranscript_mycInfect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RXN_9958_NAD_m1\n",
      "0.0\n",
      "RXN_99581\n",
      "RXN_9958_NAD_m2\n",
      "0.0\n",
      "RXN_99582\n",
      "RXN_9958_NAD_m1\n",
      "0.0\n",
      "RXN_99581\n",
      "RXN_9958_NAD_m2\n",
      "0.0\n",
      "RXN_99582\n",
      "RXN_9958_NAD_m1\n",
      "0.0\n",
      "RXN_99581\n",
      "RXN_9958_NAD_m2\n",
      "0.0\n",
      "RXN_99582\n",
      "RXN_9958_NAD_m1\n",
      "0.0\n",
      "RXN_99581\n",
      "RXN_9958_NAD_m2\n",
      "0.0\n",
      "RXN_99582\n",
      "RXN_9958_NAD_m1\n",
      "0.0\n",
      "RXN_99581\n",
      "RXN_9958_NAD_m2\n",
      "0.0\n",
      "RXN_99582\n"
     ]
    }
   ],
   "source": [
    "def generateFluxDictIgnoringCompartmentalization(model):\n",
    "    fluxes = dict()\n",
    "    for r in model.reactions:\n",
    "        fluxes[r.id] = r.flux\n",
    "    fluxes_new = dict()\n",
    "    processedRxn = set()\n",
    "    for rxn in fluxes:\n",
    "        RXN = str(rxn[:rxn.rindex(\"_\")]+rxn[len(rxn)-1]).replace(\"_NADP\",\"\").replace(\"_NAD\",\"\")\n",
    "        if RXN in  processedRxn:\n",
    "            continue\n",
    "        else:\n",
    "            processedRxn.add(RXN)\n",
    "        tot = 0\n",
    "        for rxn2 in fluxes:\n",
    "            if RXN == str(rxn2[:rxn2.rindex(\"_\")]+rxn2[len(rxn2)-1]).replace(\"_NADP\",\"\").replace(\"_NAD\",\"\"):\n",
    "                if \"RXN_9958\" in RXN:\n",
    "                    print rxn2\n",
    "                    print abs(fluxes[rxn2])\n",
    "                    print RXN\n",
    "                tot = tot+abs(fluxes[rxn2])\n",
    "        fluxes_new[RXN]=tot\n",
    "    return fluxes_new\n",
    "                \n",
    "\n",
    "\n",
    "\n",
    "modelDictNonCompartmentalized = dict()\n",
    "for i in range(1,6):\n",
    "    modelDictNonCompartmentalized[i] = generateFluxDictIgnoringCompartmentalization(modelDict[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modelDictNonCompartmentalized[1]['RXN_99582']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "PCCdict=dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import pearsonr\n",
    "\n",
    "processedRxn = set()\n",
    "for rxn in totalTranscriptDict[1].keys():\n",
    "    R = str(rxn[:rxn.rindex(\"_\")]+rxn[len(rxn)-1]).replace(\"_NADP\",\"\").replace(\"_NAD\",\"\")\n",
    "    if R in processedRxn:\n",
    "        continue\n",
    "    else:\n",
    "        processedRxn.add(R)\n",
    "    xlist = list()\n",
    "    ylist = list()\n",
    "    for i in range(1,6):\n",
    "        #print rxn\n",
    "        for j in range(0,len(totalTranscriptDict[i][rxn])):\n",
    "            xlist.append(modelDictNonCompartmentalized[i][R])\n",
    "            ylist.append(totalTranscriptDict[i][rxn][j])\n",
    "    ans = pearsonr(xlist,ylist)  \n",
    "    PCCdict[rxn]=ans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'METHYLCROTONYL_COA_CARBOXYLASE_RXN_m2': (nan, 1.0),\n",
       " 'METHYLCROTONYL_COA_CARBOXYLASE_RXN_m1': (nan, 1.0),\n",
       " 'H_mc1': (nan, 1.0),\n",
       " 'H_mc2': (nan, 1.0),\n",
       " 'Protein_Processing_c1': (nan, 1.0),\n",
       " 'Protein_Processing_c2': (nan, 1.0),\n",
       " 'Mg_biomass2': (nan, 1.0),\n",
       " 'Mg_biomass1': (nan, 1.0),\n",
       " 'AMP_pc1': (nan, 1.0),\n",
       " 'RXN_9958_NAD_m2': (nan, 1.0),\n",
       " 'RXN_6884_m1': (nan, 1.0),\n",
       " 'RXN_6884_m2': (nan, 1.0),\n",
       " 'pTYR_biomass2': (nan, 1.0),\n",
       " 'pTYR_biomass1': (nan, 1.0),\n",
       " 'MDA_Fd_Ascorbate_p2': (nan, 1.0),\n",
       " 'MDA_Fd_Ascorbate_p1': (nan, 1.0),\n",
       " '2_PERIOD_3_PERIOD_1_PERIOD_23_RXN_p2': (nan, 1.0),\n",
       " '2_PERIOD_3_PERIOD_1_PERIOD_23_RXN_p1': (nan, 1.0),\n",
       " 'RXN_6902_m1': (nan, 1.0),\n",
       " 'RXN_6902_m2': (nan, 1.0),\n",
       " 'pTHR_biomass2': (nan, 1.0),\n",
       " 'pTHR_biomass1': (nan, 1.0),\n",
       " 'ASPARAGHYD_RXN_c1': (nan, 1.0),\n",
       " 'ASPARAGHYD_RXN_c2': (nan, 1.0),\n",
       " '2KG_SUC_mc1': (nan, 1.0),\n",
       " 'HIS_ORNITHINE_mc2': (nan, 1.0),\n",
       " 'HIS_ORNITHINE_mc1': (nan, 1.0),\n",
       " 'RXN_11832_m1': (nan, 1.0),\n",
       " 'RXN_11832_m2': (nan, 1.0),\n",
       " 'GLUTDECARBOX_RXN_c1': (-0.5191227391399041, 0.0023309916567849334),\n",
       " 'GLYOHMETRANS_RXN_c1': (-0.660817486492378, 3.845116918835141e-05),\n",
       " 'RXN66_1_c2': (nan, 1.0),\n",
       " 'GLYOHMETRANS_RXN_c2': (-0.8888984348004929, 1.0871841744300075e-11),\n",
       " 'GLUTAMINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'GLUTAMINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'LYS_pc2': (nan, 1.0),\n",
       " 'LYS_pc1': (nan, 1.0),\n",
       " 'FORMATETHFLIG_RXN_p1': (nan, 1.0),\n",
       " 'FORMATETHFLIG_RXN_p2': (nan, 1.0),\n",
       " 'Photon_tx1': (nan, 1.0),\n",
       " 'FUMARYLACETOACETASE_RXN_c2': (nan, 1.0),\n",
       " 'FUMARYLACETOACETASE_RXN_c1': (nan, 1.0),\n",
       " 'Photon_tx2': (nan, 1.0),\n",
       " 'Starch_biomass1': (nan, 1.0),\n",
       " 'ACSERLY_RXN_c2': (nan, 1.0),\n",
       " 'ACSERLY_RXN_c1': (-0.45409758477113293, 0.00903639935426532),\n",
       " 'Starch_biomass2': (nan, 1.0),\n",
       " 'TRP_PROTON_vc1': (nan, 1.0),\n",
       " 'RXN_9958_NAD_m1': (nan, 1.0),\n",
       " 'TRP_PROTON_vc2': (nan, 1.0),\n",
       " '2_AMINOADIPATE_AMINOTRANSFERASE_RXN_c1': (nan, 1.0),\n",
       " '2_AMINOADIPATE_AMINOTRANSFERASE_RXN_c2': (nan, 1.0),\n",
       " 'O2_tx1': (nan, 1.0),\n",
       " 'O2_tx2': (nan, 1.0),\n",
       " 'AMP_pc2': (nan, 1.0),\n",
       " 'PRPPAMIDOTRANS_RXN_p1': (-0.7886874094378756, 8.257405040207659e-08),\n",
       " 'PRPPAMIDOTRANS_RXN_p2': (0.1924574245777729, 0.2912909091772444),\n",
       " 'DUMP_pc2': (nan, 1.0),\n",
       " 'THREDEHYD_RXN_p2': (0.3611337174894579, 0.042284055647694836),\n",
       " 'THREDEHYD_RXN_p1': (nan, 1.0),\n",
       " '3_ISOPROPYLMALDEHYDROG_RXN_p1': (-0.8852634169910315, 1.718399720870457e-11),\n",
       " '3_HYDROXYISOBUTYRYL_COA_HYDROLASE_RXN_m2': (0.5867017547065178,\n",
       "  0.0004166359359108814),\n",
       " '3_HYDROXYISOBUTYRYL_COA_HYDROLASE_RXN_m1': (nan, 1.0),\n",
       " 'INORGPYROPHOSPHAT_RXN_p1': (0.5236765889987743, 0.0020984959219726264),\n",
       " 'INORGPYROPHOSPHAT_RXN_p2': (-0.4330540781688674, 0.013295674258123145),\n",
       " 'RXN_9633_p1': (nan, 1.0),\n",
       " 'pARG_biomass1': (nan, 1.0),\n",
       " 'pARG_biomass2': (nan, 1.0),\n",
       " 'RXN_9633_p2': (nan, 1.0),\n",
       " 'FERULIC_ACID_pc1': (nan, 1.0),\n",
       " 'LYS_PROTON_vc2': (nan, 1.0),\n",
       " 'LYS_PROTON_vc1': (nan, 1.0),\n",
       " 'LEU_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'PRIBFAICARPISOM_RXN_p2': (-0.9439404241484003, 5.535570591004293e-16),\n",
       " 'PRIBFAICARPISOM_RXN_p1': (-0.8560819143617919, 4.1986677345018864e-10),\n",
       " 'sALA_biomass2': (nan, 1.0),\n",
       " 'sALA_biomass1': (nan, 1.0),\n",
       " 'ACETOVANILLONE_SYNTHASE_c1': (nan, 1.0),\n",
       " 'RXN_1841_v1': (nan, 1.0),\n",
       " 'VAL_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'GLY_pc1': (nan, 1.0),\n",
       " 'GLY_pc2': (nan, 1.0),\n",
       " 'CITSYN_RXN_x2': (-0.1919002437155611, 0.2927167509036614),\n",
       " 'RXN_15130_p2': (nan, 1.0),\n",
       " 'RXN_15130_p1': (nan, 1.0),\n",
       " 'RXN_6383_x2': (nan, 1.0),\n",
       " 'RXN_6383_x1': (nan, 1.0),\n",
       " 'RIBULP3EPIM_RXN_p1': (-0.41381791657510203, 0.01855179049499982),\n",
       " 'RIBULP3EPIM_RXN_p2': (0.4836194223427465, 0.005043805265360276),\n",
       " 'ARG_pc2': (nan, 1.0),\n",
       " 'ARG_pc1': (nan, 1.0),\n",
       " 'ASPAMINOTRANS_RXN_c1': (0.5451911373880751, 0.0012518460709565525),\n",
       " 'ACONITATEHYDR_RXN_c1': (-0.5547765676628739, 0.0009833406887284574),\n",
       " 'ACONITATEHYDR_RXN_c2': (-0.45221365491233345, 0.009363304845456706),\n",
       " 'ASPAMINOTRANS_RXN_c2': (-0.857080283676286, 3.808912594770202e-10),\n",
       " 'THREO_DS_ISO_CITRATE_xc2': (nan, 1.0),\n",
       " 'THREO_DS_ISO_CITRATE_xc1': (nan, 1.0),\n",
       " 'NADPHoxm_tx1': (nan, 1.0),\n",
       " 'PEPCARBOXYKIN_RXN_c1': (nan, 1.0),\n",
       " 'PEPCARBOXYKIN_RXN_c2': (nan, 1.0),\n",
       " '6PGLUCONDEHYDROG_RXN_c2': (nan, 1.0),\n",
       " '6PGLUCONDEHYDROG_RXN_c1': (nan, 1.0),\n",
       " 'PYRUVDEH_RXN_p2': (nan, 1.0),\n",
       " 'PYRUVDEH_RXN_p1': (nan, 1.0),\n",
       " '2KG_ACONITATE_mc1': (nan, 1.0),\n",
       " 'MANNPGUANYLTRANGDP_RXN_c2': (nan, 1.0),\n",
       " 'MANNPGUANYLTRANGDP_RXN_c1': (nan, 1.0),\n",
       " '2KG_ACONITATE_mc2': (nan, 1.0),\n",
       " 'GAP_Pi_pc1': (nan, 1.0),\n",
       " 'GAP_Pi_pc2': (nan, 1.0),\n",
       " 'sTHR_biomass1': (nan, 1.0),\n",
       " 'sTHR_biomass2': (nan, 1.0),\n",
       " 'GCVMULTI_RXN_m1': (-0.46679119094686067, 0.0070758160068118735),\n",
       " 'GCVMULTI_RXN_m2': (-0.8729538927229212, 7.276974599152532e-11),\n",
       " '1_PERIOD_10_PERIOD_2_PERIOD_2_RXN_mc2': (-0.6751905835114486,\n",
       "  2.2425682450596566e-05),\n",
       " '1_PERIOD_10_PERIOD_2_PERIOD_2_RXN_mc1': (0.20270975027102286,\n",
       "  0.26584683305156676),\n",
       " 'CYSPH_RXN_p1': (-0.31971437354413973, 0.07446069729301044),\n",
       " 'CYSPH_RXN_p2': (nan, 1.0),\n",
       " 'ADPREDUCT_RXN_p2': (nan, 1.0),\n",
       " 'ADPREDUCT_RXN_p1': (nan, 1.0),\n",
       " 'L_ASCORBATE_PEROXIDASE_RXN_m1': (-0.08472107848145302, 0.644785978493703),\n",
       " 'L_ASCORBATE_PEROXIDASE_RXN_m2': (nan, 1.0),\n",
       " 'TYR_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'TYR_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'pGLU_biomass2': (nan, 1.0),\n",
       " 'pGLU_biomass1': (nan, 1.0),\n",
       " 'ACETYL_COA_CARBOXYLTRANSFER_RXN_p1': (nan, 1.0),\n",
       " 'ACETYL_COA_CARBOXYLTRANSFER_RXN_p2': (nan, 1.0),\n",
       " 'NADH_DEHYDROG_A_RXN_mi2': (-0.7291667447491005, 2.2067282934446483e-06),\n",
       " 'NADH_DEHYDROG_A_RXN_mi1': (0.10260998591034175, 0.5762746428711668),\n",
       " '10_FORMYL_THF_pc1': (nan, 1.0),\n",
       " 'GGPP_biosynthesis_p1': (nan, 1.0),\n",
       " 'GGPP_biosynthesis_p2': (nan, 1.0),\n",
       " '10_FORMYL_THF_pc2': (nan, 1.0),\n",
       " 'ADP_pc2': (nan, 1.0),\n",
       " 'ADP_pc1': (nan, 1.0),\n",
       " 'pVAL_biomass2': (nan, 1.0),\n",
       " 'pVAL_biomass1': (nan, 1.0),\n",
       " 'Mitochondrial_ATP_Synthase_m1': (nan, 1.0),\n",
       " 'Mitochondrial_ATP_Synthase_m2': (nan, 1.0),\n",
       " 'LPG_biosynthesis_c2': (nan, 1.0),\n",
       " 'LPG_biosynthesis_c1': (nan, 1.0),\n",
       " 'GART_RXN_p1': (-0.3659236579339913, 0.03943004512043868),\n",
       " 'GART_RXN_p2': (-0.5010439966054255, 0.0034886853834862454),\n",
       " 'MALONYL_COA_ACP_TRANSACYL_RXN_p2': (0.3630576074565372,\n",
       "  0.041118446118825425),\n",
       " 'MALONYL_COA_ACP_TRANSACYL_RXN_p1': (0.3923733345508538, 0.02633595000816246),\n",
       " 'AIRCARBOXY_RXN_p2': (-0.0514504786615958, 0.7797434176114054),\n",
       " 'AIRCARBOXY_RXN_p1': (0.32179689994194804, 0.07248665328653353),\n",
       " 'THF_mc1': (nan, 1.0),\n",
       " 'GLUTACONYL_COA_DECARBOXYLASE_RXN_x1': (nan, 1.0),\n",
       " 'GLUTACONYL_COA_DECARBOXYLASE_RXN_x2': (nan, 1.0),\n",
       " 'THF_mc2': (nan, 1.0),\n",
       " 'ADENOSYLHOMOCYSTEINASE_RXN_c1': (nan, 1.0),\n",
       " 'ADENOSYLHOMOCYSTEINASE_RXN_c2': (0.07214732517470816, 0.6947617566871495),\n",
       " 'FORMATE_pc1': (nan, 1.0),\n",
       " 'FORMATE_pc2': (nan, 1.0),\n",
       " 'RXN0_5224_c2': (nan, 1.0),\n",
       " 'GLUCURONOKINASE_RXN_c1': (nan, 1.0),\n",
       " 'GLUCURONOKINASE_RXN_c2': (nan, 1.0),\n",
       " 'DUDPKIN_RXN_p1': (nan, 1.0),\n",
       " 'DUDPKIN_RXN_p2': (nan, 1.0),\n",
       " 'THR_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'THR_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'PEPDEPHOS_RXN_p1': (nan, 1.0),\n",
       " 'PEPDEPHOS_RXN_p2': (-0.7621327483694249, 4.0179515241136187e-07),\n",
       " 'ORNCARBAMTRANSFER_RXN_p1': (-0.793895595987502, 5.896814674064362e-08),\n",
       " 'L_ALPHA_ALANINE_PROTON_vc1': (nan, 1.0),\n",
       " 'SUC_ISOCITRATE_mc2': (nan, 1.0),\n",
       " 'SUC_ISOCITRATE_mc1': (nan, 1.0),\n",
       " 'GAPOXNPHOSPHN_RXN_c1': (-0.5817543254735748, 0.00047871099273216863),\n",
       " 'RXN0_5330_NAD_mi1': (nan, 1.0),\n",
       " 'ISOCITDEH_RXN_x1': (0.37770177610417816, 0.03306806678645236),\n",
       " 'ISOCITDEH_RXN_x2': (-0.5035050497587761, 0.003306554514868215),\n",
       " 'NH4_ec1': (nan, 1.0),\n",
       " '2KETO_4METHYL_PENTANOATE_DEHYDROG_RXN_m1': (nan, 1.0),\n",
       " '2KETO_4METHYL_PENTANOATE_DEHYDROG_RXN_m2': (nan, 1.0),\n",
       " 'HOMOCYSMETB12_RXN_c1': (nan, 1.0),\n",
       " '2_KETO_ADIPATE_DEHYDROG_RXN_m1': (nan, 1.0),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_9_RXN_c2': (nan, 1.0),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_9_RXN_c1': (nan, 1.0),\n",
       " 'GLUCOKIN_RXN_p1': (nan, 1.0),\n",
       " 'LEUCINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'LEUCINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'GLUCOKIN_RXN_p2': (nan, 1.0),\n",
       " 'RXN_3523_p2': (nan, 1.0),\n",
       " 'RXN_3523_p1': (nan, 1.0),\n",
       " 'RXN_9549_p1': (0.03858165042071638, 0.8339445131333396),\n",
       " 'RXN_9549_p2': (nan, 1.0),\n",
       " 'GLYCINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'CHORISMATEMUT_RXN_p1': (0.02670022796120972, 0.8846668971678491),\n",
       " 'GLUTAMIDOTRANS_RXN_p1': (-0.784624065666748, 1.0669395116707849e-07),\n",
       " 'GLUTAMIDOTRANS_RXN_p2': (-0.9292230961496135, 1.654356574205907e-14),\n",
       " '2TRANSKETO_RXN_p1': (nan, 1.0),\n",
       " 'RXN_9659_p1': (-0.028007694807174827, 0.8790595043173559),\n",
       " 'RXN_9659_p2': (-0.2797032676502659, 0.12104066029338631),\n",
       " 'HIS_PROTON_vc2': (nan, 1.0),\n",
       " 'HIS_PROTON_vc1': (nan, 1.0),\n",
       " 'ACET_pc1': (nan, 1.0),\n",
       " 'sFUM_biomass2': (nan, 1.0),\n",
       " 'sFUM_biomass1': (nan, 1.0),\n",
       " 'RXN1G_349_p2': (nan, 1.0),\n",
       " 'GLUTATHIONE_REDUCT_NADPH_RXN_p1': (-0.7382921780014017,\n",
       "  1.412267386808437e-06),\n",
       " 'HYDROXYPYRUVATE_REDUCTASE_RXN_NAD_x2': (nan, 1.0),\n",
       " 'HYDROXYPYRUVATE_REDUCTASE_RXN_NAD_x1': (-0.2703220063803485,\n",
       "  0.13456222033886095),\n",
       " '3_HYDROXBUTYRYL_COA_DEHYDRATASE_RXN_x1': (nan, 1.0),\n",
       " 'ARG_LYS_mc1': (nan, 1.0),\n",
       " '2KG_CIT_mc1': (nan, 1.0),\n",
       " 'N_ACETYLGLUTPREDUCT_RXN_p1': (0.5587146017170402, 0.0008886137982872845),\n",
       " 'N_ACETYLGLUTPREDUCT_RXN_p2': (nan, 1.0),\n",
       " 'L_ALPHA_ALANINE_v_dielTransfer': (nan, 1.0),\n",
       " 'PALMITATE_pc1': (nan, 1.0),\n",
       " 'PALMITATE_pc2': (nan, 1.0),\n",
       " 'SULFITE_REDUCTASE_FERREDOXIN_RXN_p1': (-0.31227237416045317,\n",
       "  0.0818572057089132),\n",
       " 'Phloem_output_tx2': (nan, 1.0),\n",
       " 'Phloem_output_tx1': (nan, 1.0),\n",
       " 'FGAMSYN_RXN_p2': (0.29085719417284367, 0.10630837400615588),\n",
       " 'FGAMSYN_RXN_p1': (-0.6955761294068876, 9.90519323882973e-06),\n",
       " 'RXN_9651_p1': (nan, 1.0),\n",
       " 'RXN_9651_p2': (nan, 1.0),\n",
       " 'MALSYN_RXN_x1': (-0.1291262911662983, 0.4812183648864825),\n",
       " 'PHOSGLYPHOS_RXN_p1': (nan, 1.0),\n",
       " 'PHOSGLYPHOS_RXN_p2': (nan, 1.0),\n",
       " '3PGAREARR_RXN_c1': (0.5677600350650885, 0.0007007815595027012),\n",
       " '3PGAREARR_RXN_c2': (-0.7813656222572966, 1.3052444564420525e-07),\n",
       " 'RXN_3521_p1': (nan, 1.0),\n",
       " 'OROTPDECARB_RXN_p1': (nan, 1.0),\n",
       " 'ALANINE_AMINOTRANSFERASE_RXN_c2': (-0.6514891897916363,\n",
       "  5.3749442822635145e-05),\n",
       " 'ALANINE_AMINOTRANSFERASE_RXN_c1': (-0.8424845678990908,\n",
       "  1.4781228641929706e-09),\n",
       " 'ASPARAGINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'Ferredoxin_Plastoquinone_Reductase_p2': (nan, 1.0),\n",
       " 'RXN_9540_p2': (nan, 1.0),\n",
       " 'RXN_9540_p1': (nan, 1.0),\n",
       " 'Ferredoxin_Plastoquinone_Reductase_p1': (nan, 1.0),\n",
       " 'RXN_1104_p2': (nan, 1.0),\n",
       " 'RXN_1104_p1': (-0.43071114147179634, 0.013859956946581126),\n",
       " 'DIHYDROPICRED_RXN_NADP_p2': (nan, 1.0),\n",
       " 'DIHYDROPICRED_RXN_NADP_p1': (-0.5600644952841839, 0.0008580414973980897),\n",
       " 'PRO_v_dielTransfer': (nan, 1.0),\n",
       " 'ARG_PROTON_vc2': (nan, 1.0),\n",
       " 'RXN0_2382_p1': (-0.6182264469082813, 0.00016273836942113328),\n",
       " 'SUC_SO4_mc2': (nan, 1.0),\n",
       " 'SUC_SO4_mc1': (nan, 1.0),\n",
       " 'RXN_9650_p2': (nan, 1.0),\n",
       " 'RXN_9650_p1': (nan, 1.0),\n",
       " 'Octadecadienoyl_ACP_hydrolase_p1': (nan, 1.0),\n",
       " 'Octadecadienoyl_ACP_hydrolase_p2': (nan, 1.0),\n",
       " 'ARG_ORNITHINE_mc2': (nan, 1.0),\n",
       " 'ARG_ORNITHINE_mc1': (nan, 1.0),\n",
       " 'SERINE_GLYOXYLATE_AMINOTRANSFERASE_RXN_x2': (nan, 1.0),\n",
       " 'SERINE_GLYOXYLATE_AMINOTRANSFERASE_RXN_x1': (-0.5153378006784595,\n",
       "  0.002540937206598889),\n",
       " 'ISOCITRATE_DEHYDROGENASE_NAD_RXN_m1': (0.18296847409657188,\n",
       "  0.31617904287050264),\n",
       " 'ISOCITRATE_DEHYDROGENASE_NAD_RXN_m2': (nan, 1.0),\n",
       " 'Glycerate_xc1': (nan, 1.0),\n",
       " 'Glycerate_xc2': (nan, 1.0),\n",
       " 'RXN_9653_p2': (nan, 1.0),\n",
       " 'RXN_9653_p1': (nan, 1.0),\n",
       " 'LEU_mc1': (nan, 1.0),\n",
       " 'LEU_mc2': (nan, 1.0),\n",
       " 'RXN_9657_p2': (-0.2797032676502659, 0.12104066029338631),\n",
       " 'RXN_9657_p1': (-0.028007694807174827, 0.8790595043173559),\n",
       " 'GLT_xc1': (nan, 1.0),\n",
       " 'GLT_xc2': (nan, 1.0),\n",
       " 'RXN_9654_p2': (nan, 1.0),\n",
       " 'RXN_9654_p1': (nan, 1.0),\n",
       " 'Fatty_Acids_pc1': (nan, 1.0),\n",
       " 'Fatty_Acids_pc2': (nan, 1.0),\n",
       " 'RXN_9655_p1': (nan, 1.0),\n",
       " 'RXN_9655_p2': (nan, 1.0),\n",
       " 'SO4_pc1': (nan, 1.0),\n",
       " 'SO4_pc2': (nan, 1.0),\n",
       " 'RXN_9528_p1': (nan, 1.0),\n",
       " 'RXN_9528_p2': (nan, 1.0),\n",
       " 'THR_pc1': (nan, 1.0),\n",
       " 'THR_pc2': (nan, 1.0),\n",
       " 'Cellulose_biomass1': (nan, 1.0),\n",
       " 'METHYLACYLYLCOA_HYDROXY_RXN_m2': (nan, 1.0),\n",
       " 'METHYLACYLYLCOA_HYDROXY_RXN_m1': (nan, 1.0),\n",
       " 'Cellulose_biomass2': (nan, 1.0),\n",
       " '2_METHYL_ACETO_ACETYL_COA_mc1': (nan, 1.0),\n",
       " '2_METHYL_ACETO_ACETYL_COA_mc2': (nan, 1.0),\n",
       " 'DUTP_PYROP_RXN_p1': (nan, 1.0),\n",
       " 'DUTP_PYROP_RXN_p2': (nan, 1.0),\n",
       " 'ADENYLOSUCCINATE_SYNTHASE_RXN_p1': (-0.44152456476311225,\n",
       "  0.011413616802725826),\n",
       " 'ADENYLOSUCCINATE_SYNTHASE_RXN_p2': (-0.3526228966683107, 0.0477623834282189),\n",
       " 'ACETYL_COA_ACETYLTRANSFER_RXN_x1': (nan, 1.0),\n",
       " 'ACETYL_COA_ACETYLTRANSFER_RXN_x2': (0.09076237673029434, 0.621289215899055),\n",
       " 'CO2_tx1': (nan, 1.0),\n",
       " 'MercaptoPyruvateSulfurtransferase_m2': (nan, 1.0),\n",
       " 'MercaptoPyruvateSulfurtransferase_m1': (nan, 1.0),\n",
       " 'AMP_DEAMINASE_RXN_c1': (-0.3492881507518599, 0.050057364794861656),\n",
       " 'AMP_DEAMINASE_RXN_c2': (nan, 1.0),\n",
       " 'ACETOVANILLONE_SYNTHASE_c2': (nan, 1.0),\n",
       " 'RXN0_5224_c1': (nan, 1.0),\n",
       " 'DSERDEAM_RXN_c1': (-0.1431732555857289, 0.4343748575522093),\n",
       " 'DSERDEAM_RXN_c2': (nan, 1.0),\n",
       " 'HYDROXYMETHYLGLUTARYL_COA_LYASE_RXN_m1': (nan, 1.0),\n",
       " 'HYDROXYMETHYLGLUTARYL_COA_LYASE_RXN_m2': (nan, 1.0),\n",
       " '2_PERIOD_3_PERIOD_1_PERIOD_133_RXN_c2': (nan, 1.0),\n",
       " '2_PERIOD_3_PERIOD_1_PERIOD_133_RXN_c1': (nan, 1.0),\n",
       " 'MALEYLACETOACETATE_ISOMERASE_RXN_c2': (nan, 1.0),\n",
       " 'HOMOSERKIN_RXN_p1': (-0.5618140063997592, 0.0008197967049687572),\n",
       " 'HOMOSERKIN_RXN_p2': (nan, 1.0),\n",
       " 'MALEYLACETOACETATE_ISOMERASE_RXN_c1': (nan, 1.0),\n",
       " '3_ISOPROPYLMALDEHYDROG_RXN_p2': (-0.9070848140004575, 8.435005372865012e-13),\n",
       " 'GLUC1PURIDYLTRANS_RXN_p1': (-0.7007513808844454, 7.964508831275031e-06),\n",
       " 'GLUC1PURIDYLTRANS_RXN_p2': (-0.7788715350944653, 1.5195511894006942e-07),\n",
       " 'RXN1G_445_p2': (nan, 1.0),\n",
       " 'RXN1G_445_p1': (nan, 1.0),\n",
       " 'H2O_xc1': (nan, 1.0),\n",
       " 'MAL_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'MAL_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'DUMP_pc1': (nan, 1.0),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERILEU_RXN_m2': (0.7283486431123286,\n",
       "  2.2948315228969645e-06),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERILEU_RXN_m1': (-0.6844388125698416,\n",
       "  1.5603289240221343e-05),\n",
       " 'Nitrate_tx1': (nan, 1.0),\n",
       " '1_PERIOD_1_PERIOD_1_PERIOD_178_RXN_x2': (nan, 1.0),\n",
       " '1_PERIOD_1_PERIOD_1_PERIOD_178_RXN_x1': (nan, 1.0),\n",
       " 'Nitrate_tx2': (nan, 1.0),\n",
       " 'FUM_PROTON_vc2': (nan, 1.0),\n",
       " 'FUM_PROTON_vc1': (nan, 1.0),\n",
       " 'NADPHoxc_tx2': (nan, 1.0),\n",
       " 'NADPHoxc_tx1': (nan, 1.0),\n",
       " 'HOMOCYSMETB12_RXN_p2': (nan, 1.0),\n",
       " 'METHYLENE_THF_mc2': (nan, 1.0),\n",
       " 'RXN_13202_p1': (nan, 1.0),\n",
       " 'RXN_13202_p2': (nan, 1.0),\n",
       " 'METHYLENE_THF_mc1': (nan, 1.0),\n",
       " 'SER_v_dielTransfer': (nan, 1.0),\n",
       " 'PYRUVATE_pc1': (nan, 1.0),\n",
       " 'PYRUVATE_pc2': (nan, 1.0),\n",
       " 'SHIKIMATE_KINASE_RXN_p2': (nan, 1.0),\n",
       " 'SHIKIMATE_KINASE_RXN_p1': (0.12786711642533213, 0.4855408985620753),\n",
       " 'SUPEROX_DISMUT_RXN_c2': (nan, 1.0),\n",
       " 'SUPEROX_DISMUT_RXN_c1': (nan, 1.0),\n",
       " 'pTRP_biomass2': (nan, 1.0),\n",
       " 'DIHYDROXYISOVALDEHYDRAT_RXN_p2': (nan, 1.0),\n",
       " 'RXN_9660_p1': (-0.028007694807174827, 0.8790595043173559),\n",
       " 'Pi_PROTON_mc1': (nan, 1.0),\n",
       " 'Pi_PROTON_mc2': (nan, 1.0),\n",
       " 'RXN_9520_p1': (nan, 1.0),\n",
       " 'RXN_9520_p2': (nan, 1.0),\n",
       " 'RXN_12486_c2': (nan, 1.0),\n",
       " 'RXN_6384_x1': (nan, 1.0),\n",
       " 'RXN_9635_p1': (-0.2797032676502659, 0.12104066029338631),\n",
       " 'IPP_biosynthesis_c1': (nan, 1.0),\n",
       " 'IPP_biosynthesis_c2': (nan, 1.0),\n",
       " 'ARGSUCCINLYA_RXN_p2': (nan, 1.0),\n",
       " 'pALA_biomass1': (nan, 1.0),\n",
       " 'pALA_biomass2': (nan, 1.0),\n",
       " 'ARGSUCCINLYA_RXN_p1': (0.34142843058847183, 0.05581390766208836),\n",
       " 'DIHYDROOROT_RXN_p2': (nan, 1.0),\n",
       " 'DIHYDROOROT_RXN_p1': (nan, 1.0),\n",
       " 'GLUTAMINESYN_RXN_p2': (-0.46286185883930314, 0.007639704667899129),\n",
       " 'GLUTAMINESYN_RXN_p1': (0.4374900573123, 0.012279979491029552),\n",
       " 'HOMOGENTISATE_12_DIOXYGENASE_RXN_c2': (nan, 1.0),\n",
       " 'HOMOGENTISATE_12_DIOXYGENASE_RXN_c1': (nan, 1.0),\n",
       " 'IMPCYCLOHYDROLASE_RXN_p2': (-0.6302515861201856, 0.0001106542947449623),\n",
       " 'IMPCYCLOHYDROLASE_RXN_p1': (-0.7851162004628333, 1.0346336034563059e-07),\n",
       " 'RXN_7978_p2': (nan, 1.0),\n",
       " 'RXN_7978_p1': (nan, 1.0),\n",
       " 'TYROSINE_AMINOTRANSFERASE_RXN_p2': (nan, 1.0),\n",
       " 'TYROSINE_AMINOTRANSFERASE_RXN_p1': (nan, 1.0),\n",
       " 'METHENYLTHFCYCLOHYDRO_RXN_c1': (-0.3446719931238632, 0.05337799678296207),\n",
       " 'RXN_9623_p2': (nan, 1.0),\n",
       " 'RXN_11213_m2': (nan, 1.0),\n",
       " 'GLN_PROTON_vc2': (nan, 1.0),\n",
       " 'GLYC3PDEHYDROGBIOSYN_RXN_c2': (nan, 1.0),\n",
       " '2_PERIOD_4_PERIOD_1_PERIOD_46_RXN_p1': (nan, 1.0),\n",
       " '2_PERIOD_4_PERIOD_1_PERIOD_46_RXN_p2': (nan, 1.0),\n",
       " 'GLYC3PDEHYDROGBIOSYN_RXN_c1': (nan, 1.0),\n",
       " 'sGLN_biomass1': (nan, 1.0),\n",
       " 'sGLN_biomass2': (nan, 1.0),\n",
       " 'PREPHENATE_TRANSAMINE_RXN_p1': (-0.6954646241335706, 9.951350006477724e-06),\n",
       " 'PRO_mc2': (nan, 1.0),\n",
       " 'PRO_mc1': (nan, 1.0),\n",
       " 'PREPHENATE_TRANSAMINE_RXN_p2': (0.8086298942596717, 2.154975933959805e-08),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_25_RXN_m2': (0.7050814285643691,\n",
       "  6.612931802989648e-06),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_25_RXN_m1': (nan, 1.0),\n",
       " 'GLYCERATE_GLYCOLLATE_pc2': (nan, 1.0),\n",
       " 'GLYCERATE_GLYCOLLATE_pc1': (nan, 1.0),\n",
       " 'IMP_DEHYDROG_RXN_c1': (-0.30382735505080033, 0.09092160758839185),\n",
       " 'HIS_pc2': (nan, 1.0),\n",
       " 'HIS_pc1': (nan, 1.0),\n",
       " 'IMP_DEHYDROG_RXN_c2': (nan, 1.0),\n",
       " 'HISTIDPHOS_RXN_p2': (-0.9134852546983495, 3.020880742661018e-13),\n",
       " 'HISTIDPHOS_RXN_p1': (-0.7190694170577252, 3.5437844519808615e-06),\n",
       " 'PHOSPHOLIPASE_A2_RXN_c1': (nan, 1.0),\n",
       " 'PHOSPHOLIPASE_A2_RXN_c2': (nan, 1.0),\n",
       " 'SUC_FUM_mc1': (nan, 1.0),\n",
       " 'SUC_FUM_mc2': (nan, 1.0),\n",
       " 'UREA_mc2': (nan, 1.0),\n",
       " 'DHAP_3PGA_pc1': (nan, 1.0),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERVAL_RXN_m1': (-0.6716140952352478,\n",
       "  2.5715494914158108e-05),\n",
       " 'SUCROSE_SYNTHASE_RXN_c1': (-0.8000271482894776, 3.9180070945441705e-08),\n",
       " 'SUCROSE_SYNTHASE_RXN_c2': (-0.7774127731004427, 1.659365228997339e-07),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERVAL_RXN_m2': (-0.5197600841369852,\n",
       "  0.0022971630998724693),\n",
       " '6PFRUCTPHOS_RXN_c2': (-0.7936313032980635, 5.999806513701933e-08),\n",
       " '6PFRUCTPHOS_RXN_c1': (nan, 1.0),\n",
       " 'RXN_1781_v1': (nan, 1.0),\n",
       " 'RXN490_3650_p1': (nan, 1.0),\n",
       " 'sGABA_biomass2': (nan, 1.0),\n",
       " 'sGABA_biomass1': (nan, 1.0),\n",
       " 'CIT_xc2': (nan, 1.0),\n",
       " 'CIT_xc1': (nan, 1.0),\n",
       " 'gadoleic_acid_biosynthesis_p1': (nan, 1.0),\n",
       " 'gadoleic_acid_biosynthesis_p2': (nan, 1.0),\n",
       " 'GDPREDUCT_RXN_p1': (nan, 1.0),\n",
       " 'GDPREDUCT_RXN_p2': (nan, 1.0),\n",
       " 'PHE_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'PHE_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'MANNITOL_1_PHOSPHATASE_RXN_c1': (nan, 1.0),\n",
       " 'MANNITOL_1_PHOSPHATASE_RXN_c2': (nan, 1.0),\n",
       " 'sGLU_biomass1': (nan, 1.0),\n",
       " 'sGLU_biomass2': (nan, 1.0),\n",
       " 'ASN_v_dielTransfer': (nan, 1.0),\n",
       " 'RXN_1381_p2': (nan, 1.0),\n",
       " 'RXN_1381_p1': (nan, 1.0),\n",
       " 'AICARSYN_RXN_p1': (-0.18210440100659941, 0.3185092459314527),\n",
       " 'AICARSYN_RXN_p2': (-0.6701630688490441, 2.7169825249464858e-05),\n",
       " 'ACETOOHBUTSYN_RXN_p2': (nan, 1.0),\n",
       " 'K_ec2': (nan, 1.0),\n",
       " 'K_ec1': (nan, 1.0),\n",
       " 'ACETOOHBUTSYN_RXN_p1': (nan, 1.0),\n",
       " 'RXN_8001_p2': (-0.913660100793923, 2.9340857565428683e-13),\n",
       " 'RXN_8001_p1': (-0.7496213125452771, 7.906854193265499e-07),\n",
       " 'DIACYLGLYCEROL_O_ACYLTRANSFERASE_RXN_p2': (nan, 1.0),\n",
       " 'DIACYLGLYCEROL_O_ACYLTRANSFERASE_RXN_p1': (nan, 1.0),\n",
       " 'GLT_PROTON_vc2': (nan, 1.0),\n",
       " 'GLT_PROTON_vc1': (nan, 1.0),\n",
       " 'RXN66_3_m1': (-0.5480311969557085, 0.001166300826409358),\n",
       " 'RXN66_3_m2': (nan, 1.0),\n",
       " 'sSUCROSE_biomass1': (nan, 1.0),\n",
       " 'hexadecadienoate_ACP_desaturase_p1': (nan, 1.0),\n",
       " 'hexadecadienoate_ACP_desaturase_p2': (nan, 1.0),\n",
       " 'sSUCROSE_biomass2': (nan, 1.0),\n",
       " 'DIHYDROXYISOVALDEHYDRAT_RXN_p1': (-0.7388478565897566,\n",
       "  1.3735966528726887e-06),\n",
       " 'MAL_ISOCITRATE_mc1': (nan, 1.0),\n",
       " 'ACET_xc2': (nan, 1.0),\n",
       " 'MAL_ISOCITRATE_mc2': (nan, 1.0),\n",
       " 'lycopene_biosynthesis_p2': (nan, 1.0),\n",
       " 'lycopene_biosynthesis_p1': (nan, 1.0),\n",
       " 'pTRP_biomass1': (nan, 1.0),\n",
       " '4_AMINO_BUTYRATE_PROTON_rev_vc2': (nan, 1.0),\n",
       " '4_AMINO_BUTYRATE_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'RXN_6883_mi2': (nan, 1.0),\n",
       " 'ASPARTATEKIN_RXN_p1': (-0.7633308693890106, 3.7576931411207885e-07),\n",
       " 'ASPARTATEKIN_RXN_p2': (nan, 1.0),\n",
       " 'RXN_7677_p2': (nan, 1.0),\n",
       " 'RXN_7677_p1': (0.061856894133544935, 0.736632569089071),\n",
       " 'pHIS_biomass1': (nan, 1.0),\n",
       " '2_METHYL_3_HYDROXY_BUTYRYL_COA_xc2': (nan, 1.0),\n",
       " '2_METHYL_3_HYDROXY_BUTYRYL_COA_xc1': (nan, 1.0),\n",
       " 'RXN0_5260_m1': (nan, 1.0),\n",
       " 'RXN0_5260_m2': (nan, 1.0),\n",
       " '3_PERIOD_1_PERIOD_2_PERIOD_14_RXN_p1': (nan, 1.0),\n",
       " 'GLYCINE_AMINOTRANSFERASE_RXN_x1': (-0.4378457780643774,\n",
       "  0.012201435665323373),\n",
       " 'RXN_7676_p1': (0.06185689413354501, 0.736632569089071),\n",
       " 'RXN_7676_p2': (nan, 1.0),\n",
       " 'sMAL_biomass1': (nan, 1.0),\n",
       " 'sMAL_biomass2': (nan, 1.0),\n",
       " 'pGLY_biomass2': (nan, 1.0),\n",
       " 'pGLY_biomass1': (nan, 1.0),\n",
       " 'RXN_7674_p2': (nan, 1.0),\n",
       " 'RXN_7674_p1': (nan, 1.0),\n",
       " '3_PERIOD_2_PERIOD_1_PERIOD_48_RXN_v1': (nan, 1.0),\n",
       " 'VANILLICACID_BIOSYNTHESIS_c1': (nan, 1.0),\n",
       " 'VANILLICACID_BIOSYNTHESIS_c2': (nan, 1.0),\n",
       " '3_PERIOD_2_PERIOD_1_PERIOD_48_RXN_v2': (nan, 1.0),\n",
       " '4_AMINO_BUTYRATE_PROTON_vc1': (nan, 1.0),\n",
       " '4_AMINO_BUTYRATE_PROTON_vc2': (nan, 1.0),\n",
       " 'ILE_pc2': (nan, 1.0),\n",
       " 'ILE_pc1': (nan, 1.0),\n",
       " 'GMP_SYN_GLUT_RXN_c2': (nan, 1.0),\n",
       " 'GMP_SYN_GLUT_RXN_c1': (0.2951495561736218, 0.1010137967726792),\n",
       " 'PHE_v_dielTransfer': (nan, 1.0),\n",
       " 'RXN_7985_p2': (nan, 1.0),\n",
       " 'RXN_7985_p1': (nan, 1.0),\n",
       " 'UDP_GLUCURONATE_DECARBOXYLASE_RXN_c1': (nan, 1.0),\n",
       " 'sCIT_biomass2': (nan, 1.0),\n",
       " 'sCIT_biomass1': (nan, 1.0),\n",
       " 'UDP_GLUCURONATE_DECARBOXYLASE_RXN_c2': (nan, 1.0),\n",
       " 'RXN_7679_p1': (nan, 1.0),\n",
       " 'RXN_7679_p2': (nan, 1.0),\n",
       " 'GUANYL_KIN_RXN_c2': (nan, 1.0),\n",
       " 'GUANYL_KIN_RXN_c1': (0.31262580855390837, 0.0814936141848808),\n",
       " 'RXN_6384_x2': (nan, 1.0),\n",
       " 'RXN_12486_c1': (nan, 1.0),\n",
       " 'RXN_8871_c1': (nan, 1.0),\n",
       " 'RXN_8871_c2': (nan, 1.0),\n",
       " 'VAL_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'UDPKIN_RXN_p2': (-0.9172186800089519, 1.598902466563797e-13),\n",
       " 'SYRINGICACID_BIOSYNTHESIS_c2': (nan, 1.0),\n",
       " 'ACETOLACTSYN_RXN_p1': (-0.1323398740208633, 0.4702774326791379),\n",
       " 'MALSYN_RXN_x2': (nan, 1.0),\n",
       " 'UDPKIN_RXN_p1': (-0.7657167319435346, 3.2847400284184613e-07),\n",
       " 'SUCROSE_PHOSPHATE_SYNTHASE_RXN_c1': (-0.0970888166733911,\n",
       "  0.5970706206576805),\n",
       " 'SUCROSE_PHOSPHATE_SYNTHASE_RXN_c2': (0.0002987531923636657,\n",
       "  0.9987052240869476),\n",
       " 'GLT_MAL_pc2': (nan, 1.0),\n",
       " 'GLT_MAL_pc1': (nan, 1.0),\n",
       " 'Beta_Oxidation_x1': (nan, 1.0),\n",
       " 'Beta_Oxidation_x2': (nan, 1.0),\n",
       " 'MET_PROTON_rev_vc2': (nan, 1.0),\n",
       " '2K_ADIPATE_mc2': (nan, 1.0),\n",
       " '2K_ADIPATE_mc1': (nan, 1.0),\n",
       " 'MET_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'GLY_v_dielTransfer': (nan, 1.0),\n",
       " 'NITRATE_vc2': (nan, 1.0),\n",
       " 'NITRATE_vc1': (nan, 1.0),\n",
       " 'FERREDOXIN_NITRITE_REDUCTASE_RXN_p2': (nan, 1.0),\n",
       " '1_PERIOD_5_PERIOD_1_PERIOD_8_RXN_c1': (nan, 1.0),\n",
       " '1_PERIOD_5_PERIOD_1_PERIOD_8_RXN_c2': (-0.40914232459241495,\n",
       "  0.02006144933306155),\n",
       " 'FERREDOXIN_NITRITE_REDUCTASE_RXN_p1': (-0.29463317163417707,\n",
       "  0.1016399878043717),\n",
       " 'NITRATE_v_dielTransfer': (nan, 1.0),\n",
       " 'GMP_pc2': (nan, 1.0),\n",
       " 'GMP_pc1': (nan, 1.0),\n",
       " 'CIT_ISOCITRATE_mc2': (nan, 1.0),\n",
       " 'CIT_ISOCITRATE_mc1': (nan, 1.0),\n",
       " 'GLC_pc2': (nan, 1.0),\n",
       " 'GLC_pc1': (nan, 1.0),\n",
       " 'TYR_PROTON_vc1': (nan, 1.0),\n",
       " 'TYR_PROTON_vc2': (nan, 1.0),\n",
       " 'GLU6PDEHYDROG_RXN_p1': (nan, 1.0),\n",
       " 'ORNITHINE_GLU_AMINOTRANSFORASE_RXN_m2': (nan, 1.0),\n",
       " 'DIHYDROFOLATEREDUCT_RXN_m1': (nan, 1.0),\n",
       " 'DIHYDROFOLATEREDUCT_RXN_m2': (-0.1256572969275465, 0.49317476793094883),\n",
       " 'HISTAMINOTRANS_RXN_p1': (-0.4501726581525776, 0.009728702600164767),\n",
       " 'HISTAMINOTRANS_RXN_p2': (-0.7461157420069634, 9.491727027552207e-07),\n",
       " 'DIHYDROXYMETVALDEHYDRAT_RXN_p1': (nan, 1.0),\n",
       " 'DIHYDROXYMETVALDEHYDRAT_RXN_p2': (nan, 1.0),\n",
       " 'RXN0_5184_c1': (nan, 1.0),\n",
       " 'RXN0_5184_c2': (-0.6749178612789734, 2.2662471149173316e-05),\n",
       " 'MEPROPCOA_FAD_RXN_m2': (nan, 1.0),\n",
       " 'MEPROPCOA_FAD_RXN_m1': (nan, 1.0),\n",
       " 'GLUTAMATE_SYNTHASE_FERREDOXIN_RXN_p1': (-0.5020871577564023,\n",
       "  0.0034104556256736003),\n",
       " 'GLUTAMATE_SYNTHASE_FERREDOXIN_RXN_p2': (nan, 1.0),\n",
       " 'PHE_PROTON_vc1': (nan, 1.0),\n",
       " 'PHE_PROTON_vc2': (nan, 1.0),\n",
       " 'RIB5PISOM_RXN_p1': (-0.620463185469249, 0.00015165287528878793),\n",
       " 'ORNITHINE_CITRULLINE_mc2': (nan, 1.0),\n",
       " 'ORNITHINE_CITRULLINE_mc1': (nan, 1.0),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_2_RXN_p2': (nan, 1.0),\n",
       " '1_PERIOD_2_PERIOD_1_PERIOD_2_RXN_p1': (nan, 1.0),\n",
       " 'Plastoquinol_Oxidase_p1': (nan, 1.0),\n",
       " 'Plastoquinol_Oxidase_p2': (nan, 1.0),\n",
       " 'VAL_PROTON_vc1': (nan, 1.0),\n",
       " 'VAL_PROTON_vc2': (nan, 1.0),\n",
       " 'GLUTATHIONE_REDUCT_NADPH_RXN_p2': (nan, 1.0),\n",
       " 'NH4_tx2': (nan, 1.0),\n",
       " 'SUCCINATE_DEHYDROGENASE_UBIQUINONE_RXN_mc1': (nan, 1.0),\n",
       " 'SUCCINATE_DEHYDROGENASE_UBIQUINONE_RXN_mc2': (-0.37636629906644636,\n",
       "  0.033744536034857095),\n",
       " 'DTMP_mc2': (nan, 1.0),\n",
       " 'DTMP_mc1': (nan, 1.0),\n",
       " 'CHORISMATEMUT_RXN_p2': (0.2972923337385764, 0.09844648642970709),\n",
       " 'GLYCINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'sSER_biomass1': (nan, 1.0),\n",
       " 'sSER_biomass2': (nan, 1.0),\n",
       " 'ARG_LYS_mc2': (nan, 1.0),\n",
       " '3_HYDROXBUTYRYL_COA_DEHYDRATASE_RXN_x2': (-0.09304134346825738,\n",
       "  0.6125182455105309),\n",
       " 'CYS_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'ACETYLGLUTKIN_RXN_p1': (0.4301602380596847, 0.01399551149698864),\n",
       " 'ACETYLGLUTKIN_RXN_p2': (nan, 1.0),\n",
       " 'RXN_1121_c2': (nan, 1.0),\n",
       " 'RXN_1121_c1': (nan, 1.0),\n",
       " 'RXN_961_p1': (nan, 1.0),\n",
       " 'RXN_961_p2': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_8_PERIOD_11_RXN_r2': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_8_PERIOD_11_RXN_r1': (nan, 1.0),\n",
       " 'sILE_biomass2': (nan, 1.0),\n",
       " 'sILE_biomass1': (nan, 1.0),\n",
       " 'ARGININE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'ARGININE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'MET_PROTON_vc1': (nan, 1.0),\n",
       " 'MET_PROTON_vc2': (nan, 1.0),\n",
       " 'DIAMINOPIMDECARB_RXN_p2': (nan, 1.0),\n",
       " 'DIAMINOPIMDECARB_RXN_p1': (-0.39588992571355514, 0.024901844776114422),\n",
       " '2_PERIOD_7_PERIOD_7_PERIOD_14_RXN_m2': (nan, 1.0),\n",
       " 'Xylan_biomass1': (nan, 1.0),\n",
       " 'Xylan_biomass2': (nan, 1.0),\n",
       " 'RXN_1225_p2': (nan, 1.0),\n",
       " 'RXN_1225_p1': (nan, 1.0),\n",
       " 'RXN_1103_c2': (nan, 1.0),\n",
       " 'RXN_1103_c1': (nan, 1.0),\n",
       " 'TYR_v_dielTransfer': (nan, 1.0),\n",
       " 'PHOSPHAGLYPSYN_RXN_m1': (nan, 1.0),\n",
       " 'RXN_3522_p1': (nan, 1.0),\n",
       " 'RXN_3522_p2': (nan, 1.0),\n",
       " 'GLUC1PADENYLTRANS_RXN_p1': (-0.7654600868426799, 3.332865812408663e-07),\n",
       " 'GLUC1PADENYLTRANS_RXN_p2': (nan, 1.0),\n",
       " 'SER_PROTON_vc1': (nan, 1.0),\n",
       " 'SER_PROTON_vc2': (nan, 1.0),\n",
       " 'GLN_GLU_mc1': (nan, 1.0),\n",
       " 'GLN_GLU_mc2': (nan, 1.0),\n",
       " 'ALCOHOL_DEHYDROG_RXN_c1': (nan, 1.0),\n",
       " 'ALCOHOL_DEHYDROG_RXN_c2': (nan, 1.0),\n",
       " 'Biomass_tx2': (nan, 1.0),\n",
       " 'CYSTATHIONINE_BETA_LYASE_RXN_p2': (nan, 1.0),\n",
       " 'CYSTATHIONINE_BETA_LYASE_RXN_p1': (-0.06130096418689048, 0.7389171299457752),\n",
       " 'RXN_7903_p1': (nan, 1.0),\n",
       " 'RXN_7903_p2': (nan, 1.0),\n",
       " 'SAICARSYN_RXN_p2': (0.7077542677565938, 5.886080742164395e-06),\n",
       " 'SAICARSYN_RXN_p1': (-0.20348962224586034, 0.26397277021519794),\n",
       " 'METHENYLTHFCYCLOHYDRO_RXN_c2': (-0.5369167473295714, 0.0015331718856542666),\n",
       " 'GLN_PROTON_vc1': (nan, 1.0),\n",
       " 'UGD_RXN_c2': (nan, 1.0),\n",
       " 'UGD_RXN_c1': (nan, 1.0),\n",
       " 'RXN_11213_m1': (nan, 1.0),\n",
       " 'pASP_biomass1': (nan, 1.0),\n",
       " 'pASP_biomass2': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_1_PERIOD_90_RXN_c1': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_1_PERIOD_90_RXN_c2': (-0.9334946620478026,\n",
       "  6.6935998010794084e-15),\n",
       " 'OAA_MAL_pc1': (nan, 1.0),\n",
       " 'OAA_MAL_pc2': (nan, 1.0),\n",
       " 'PYRROLINECARBREDUCT_RXN_NAD_c1': (-0.0023633807440506713,\n",
       "  0.9897575321363832),\n",
       " 'PYRROLINECARBREDUCT_RXN_NAD_c2': (nan, 1.0),\n",
       " 'RXN_9658_p2': (-0.2797032676502659, 0.12104066029338631),\n",
       " 'RXN_9658_p1': (-0.028007694807174827, 0.8790595043173559),\n",
       " 'RXN_5682_p1': (nan, 1.0),\n",
       " 'RXN_5682_p2': (nan, 1.0),\n",
       " 'RIBULOSE_BISPHOSPHATE_CARBOXYLASE_RXN_p1': (0.11465989978257314,\n",
       "  0.5320531701089601),\n",
       " 'PHOSPHORYL_CHOLINE_mc2': (nan, 1.0),\n",
       " 'ATP_AMP_mc2': (nan, 1.0),\n",
       " 'PHOSPHORYL_CHOLINE_mc1': (nan, 1.0),\n",
       " 'RXN_7737_p1': (-0.1623050684322632, 0.37480303003162085),\n",
       " 'Glycerol_biomass2': (nan, 1.0),\n",
       " 'Glycerol_biomass1': (nan, 1.0),\n",
       " 'RXN_7737_p2': (nan, 1.0),\n",
       " 'GLY_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'GLY_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'ISOCIT_CLEAV_RXN_x2': (nan, 1.0),\n",
       " 'GLN_v_dielTransfer': (nan, 1.0),\n",
       " 'MYO_INOSITOL_1_PHOSPHATE_SYNTHASE_RXN_c2': (nan, 1.0),\n",
       " 'MYO_INOSITOL_1_PHOSPHATE_SYNTHASE_RXN_c1': (nan, 1.0),\n",
       " 'ISOCIT_CLEAV_RXN_x1': (-0.0013491432972277308, 0.9941529524110087),\n",
       " 'ISOBUTYRYL_COA_xc2': (nan, 1.0),\n",
       " 'ISOBUTYRYL_COA_xc1': (nan, 1.0),\n",
       " 'RXN_9548_p2': (nan, 1.0),\n",
       " 'RXN_9548_p1': (nan, 1.0),\n",
       " 'CDPDIGLYSYN_RXN_p1': (nan, 1.0),\n",
       " 'CDPDIGLYSYN_RXN_p2': (nan, 1.0),\n",
       " 'FattyAcid_biomass2': (nan, 1.0),\n",
       " 'FattyAcid_biomass1': (nan, 1.0),\n",
       " 'AICARTRANSFORM_RXN_p1': (-0.7851162004628333, 1.0346336034563059e-07),\n",
       " 'AICARTRANSFORM_RXN_p2': (-0.6302515861201856, 0.0001106542947449623),\n",
       " 'hexadecatrienoate_biosynthesis_p2': (nan, 1.0),\n",
       " 'hexadecatrienoate_biosynthesis_p1': (nan, 1.0),\n",
       " 'PGPPHOSPHA_RXN_p1': (nan, 1.0),\n",
       " 'PGPPHOSPHA_RXN_p2': (nan, 1.0),\n",
       " 'MAL_v_dielTransfer': (nan, 1.0),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERLEU_RXN_m1': (-0.14430209275587216,\n",
       "  0.43072245260092756),\n",
       " 'BRANCHED_CHAINAMINOTRANSFERLEU_RXN_m2': (0.05771551978272384,\n",
       "  0.7537031048800825),\n",
       " 'sLEU_biomass1': (nan, 1.0),\n",
       " 'RXN0_1483_c2': (nan, 1.0),\n",
       " 'RXN0_1483_c1': (nan, 1.0),\n",
       " 'sLEU_biomass2': (nan, 1.0),\n",
       " 'F16BDEPHOS_RXN_p1': (-0.6258954157905647, 0.00012748397739299248),\n",
       " 'F16BDEPHOS_RXN_p2': (nan, 1.0),\n",
       " 'LYS_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'FUMHYDR_RXN_m1': (-0.7400082546841971, 1.29593391650412e-06),\n",
       " 'FUMHYDR_RXN_m2': (-0.726076004068616, 2.5566232140779066e-06),\n",
       " 'LYS_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'IMIDPHOSDEHYD_RXN_p2': (-0.8598556103753544, 2.8939572126986355e-10),\n",
       " 'IMIDPHOSDEHYD_RXN_p1': (-0.646731186960318, 6.348920492536697e-05),\n",
       " 'RXN_9652_p1': (nan, 1.0),\n",
       " 'RXN_9652_p2': (nan, 1.0),\n",
       " 'Protein_Polymerisation_c1': (nan, 1.0),\n",
       " 'Protein_Polymerisation_c2': (nan, 1.0),\n",
       " 'L_ASPARTATE_GLU_mc2': (nan, 1.0),\n",
       " 'THR_PROTON_vc1': (nan, 1.0),\n",
       " 'THR_PROTON_vc2': (nan, 1.0),\n",
       " 'SER_mc1': (nan, 1.0),\n",
       " 'SER_mc2': (nan, 1.0),\n",
       " '2KG_SUC_mc2': (nan, 1.0),\n",
       " 'UDPREDUCT_RXN_p2': (nan, 1.0),\n",
       " 'UDPREDUCT_RXN_p1': (nan, 1.0),\n",
       " 'ALLYSINE_DEHYDROG_RXN_c1': (nan, 1.0),\n",
       " 'ALLYSINE_DEHYDROG_RXN_c2': (nan, 1.0),\n",
       " 'RXN_9632_p2': (nan, 1.0),\n",
       " 'RXN_9632_p1': (nan, 1.0),\n",
       " 'PROPCOASYN_RXN_x1': (nan, 1.0),\n",
       " 'PROPCOASYN_RXN_x2': (nan, 1.0),\n",
       " 'ACETOSYRINGONE_BIOSYNTHESIS_c1': (nan, 1.0),\n",
       " 'ACETOSYRINGONE_BIOSYNTHESIS_c2': (nan, 1.0),\n",
       " 'RXN_9531_p2': (nan, 1.0),\n",
       " 'RXN_9531_p1': (nan, 1.0),\n",
       " 'PSERTRANSAM_RXN_p1': (nan, 1.0),\n",
       " 'PSERTRANSAM_RXN_p2': (0.5269442767767181, 0.0019443470528710902),\n",
       " 'CDPREDUCT_RXN_p1': (nan, 1.0),\n",
       " 'CDPREDUCT_RXN_p2': (nan, 1.0),\n",
       " 'SUC_xc1': (nan, 1.0),\n",
       " 'SUC_xc2': (nan, 1.0),\n",
       " '2KG_MAL_pc1': (nan, 1.0),\n",
       " 'HOMOSERDEHYDROG_RXN_NADP_p2': (nan, 1.0),\n",
       " 'HOMOSERDEHYDROG_RXN_NADP_p1': (-0.7456111116149027, 9.742308493114063e-07),\n",
       " '1TRANSKETO_RXN_p2': (-0.5482626688560506, 0.00115956000739792),\n",
       " '1TRANSKETO_RXN_p1': (0.036781775288167305, 0.8415913823060375),\n",
       " 'Biomass_tx1': (nan, 1.0),\n",
       " 'RXN_9537_p2': (nan, 1.0),\n",
       " 'PGLUCISOM_RXN_p2': (-0.7147542203318837, 4.312513208176001e-06),\n",
       " 'PGLUCISOM_RXN_p1': (-0.7145356317303749, 4.3552053764867046e-06),\n",
       " 'ARGSUCCINSYN_RXN_p2': (nan, 1.0),\n",
       " 'ARGSUCCINSYN_RXN_p1': (-0.8676029467460734, 1.3017474413293589e-10),\n",
       " 'ASPARTATE_SEMIALDEHYDE_DEHYDROGENASE_RXN_p2': (nan, 1.0),\n",
       " 'ASPARTATE_SEMIALDEHYDE_DEHYDROGENASE_RXN_p1': (-0.7857437158923225,\n",
       "  9.947407980618898e-08),\n",
       " 'ILE_v_dielTransfer': (nan, 1.0),\n",
       " 'HS_pc1': (nan, 1.0),\n",
       " 'HS_pc2': (nan, 1.0),\n",
       " 'RXN_9535_p2': (nan, 1.0),\n",
       " 'RXN_9535_p1': (nan, 1.0),\n",
       " '3PGA_Pi_pc2': (nan, 1.0),\n",
       " 'GLY_PROTON_vc1': (nan, 1.0),\n",
       " 'GLUTACONYL_COA_mc2': (nan, 1.0),\n",
       " 'GLUTACONYL_COA_mc1': (nan, 1.0),\n",
       " 'GLY_PROTON_vc2': (nan, 1.0),\n",
       " 'SYRINGALDEHYDE_BIOSYNTHESIS_c1': (nan, 1.0),\n",
       " 'SYRINGALDEHYDE_BIOSYNTHESIS_c2': (nan, 1.0),\n",
       " 'DHAP_3PGA_pc2': (nan, 1.0),\n",
       " 'UREA_mc1': (nan, 1.0),\n",
       " 'ALANINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'ALANINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'THR_v_dielTransfer': (nan, 1.0),\n",
       " 'PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_c1': (-0.5016076645907863,\n",
       "  0.0034462241227592822),\n",
       " 'PYRUVATEORTHOPHOSPHATE_DIKINASE_RXN_c2': (nan, 1.0),\n",
       " 'MAL_CIT_rev_vc1': (nan, 1.0),\n",
       " 'MAL_CIT_rev_vc2': (nan, 1.0),\n",
       " 'RXN_10773_c1': (nan, 1.0),\n",
       " '1_PERIOD_5_PERIOD_1_PERIOD_9_RXN_c2': (-0.40914232459241495,\n",
       "  0.02006144933306155),\n",
       " '1_PERIOD_5_PERIOD_1_PERIOD_9_RXN_c1': (nan, 1.0),\n",
       " 'RXN_10773_c2': (nan, 1.0),\n",
       " 'CATAL_RXN_x1': (-0.5401370059875392, 0.0014177295298990163),\n",
       " 'SEDOBISALDOL_RXN_p1': (nan, 1.0),\n",
       " 'SUC_Pi_mc2': (nan, 1.0),\n",
       " 'CATAL_RXN_x2': (nan, 1.0),\n",
       " 'PRO_GLU_mc1': (nan, 1.0),\n",
       " 'PRO_GLU_mc2': (nan, 1.0),\n",
       " 'PGLYCDEHYDROG_RXN_p2': (-0.2511197272855026, 0.16563627757013574),\n",
       " 'PGLYCDEHYDROG_RXN_p1': (nan, 1.0),\n",
       " 'MALATE_DEH_RXN_p1': (-0.5520935071074946, 0.001052852114310601),\n",
       " 'sASP_biomass2': (nan, 1.0),\n",
       " 'sASP_biomass1': (nan, 1.0),\n",
       " 'MALATE_DEH_RXN_p2': (-0.4190211608122825, 0.01698385752959095),\n",
       " 'S_ADENMETSYN_RXN_c2': (nan, 1.0),\n",
       " 'sTYR_biomass1': (nan, 1.0),\n",
       " 'sTYR_biomass2': (nan, 1.0),\n",
       " 'S_ADENMETSYN_RXN_c1': (nan, 1.0),\n",
       " 'RXN_1781_v2': (nan, 1.0),\n",
       " 'ASPARTATE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'ASPARTATE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'CELLULOSE_SYNTHASE_GDP_FORMING_RXN_c1': (nan, 1.0),\n",
       " 'CELLULOSE_SYNTHASE_GDP_FORMING_RXN_c2': (nan, 1.0),\n",
       " 'RXN_2621_c1': (nan, 1.0),\n",
       " 'RXN_2621_c2': (nan, 1.0),\n",
       " '2KG_ISOCITRATE_mc2': (nan, 1.0),\n",
       " '2KG_ISOCITRATE_mc1': (nan, 1.0),\n",
       " 'ACONITATEDEHYDR_RXN_c1': (-0.5547765676628729, 0.0009833406887284825),\n",
       " 'LYS_ORNITHINE_mc1': (nan, 1.0),\n",
       " 'ACONITATEDEHYDR_RXN_c2': (-0.45221365491233345, 0.009363304845456706),\n",
       " 'ADENYL_KIN_RXN_m1': (0.6451820387978505, 6.698591909322532e-05),\n",
       " 'ADENYL_KIN_RXN_m2': (-0.493737878210667, 0.004081279442494503),\n",
       " 'SUCROSE_PROTON_vc1': (nan, 1.0),\n",
       " 'HMBPP_synthesis_p2': (nan, 1.0),\n",
       " 'HMBPP_synthesis_p1': (nan, 1.0),\n",
       " 'SUCROSE_PROTON_vc2': (nan, 1.0),\n",
       " 'SHIKIMATE_5_DEHYDROGENASE_RXN_p2': (nan, 1.0),\n",
       " 'SHIKIMATE_5_DEHYDROGENASE_RXN_p1': (-0.7348452012822128,\n",
       "  1.6751392556310074e-06),\n",
       " 'sARG_biomass2': (nan, 1.0),\n",
       " 'sARG_biomass1': (nan, 1.0),\n",
       " 'ACETOVANILLONE_HYDROXYLASE_c1': (nan, 1.0),\n",
       " 'K_rev_vc2': (nan, 1.0),\n",
       " 'sCYS_biomass1': (nan, 1.0),\n",
       " 'OAA_xc2': (nan, 1.0),\n",
       " 'OAA_xc1': (nan, 1.0),\n",
       " 'PHOSPHOGLUCMUT_RXN_c2': (-0.7070008788444824, 6.083248726713506e-06),\n",
       " 'PHOSPHOGLUCMUT_RXN_c1': (-0.6853653545724282, 1.5035681728927685e-05),\n",
       " 'pASN_biomass2': (nan, 1.0),\n",
       " 'Arachidoyl_ACP_desaturase_p1': (nan, 1.0),\n",
       " 'pMET_biomass2': (nan, 1.0),\n",
       " 'pMET_biomass1': (nan, 1.0),\n",
       " 'SUC_ACONITATE_mc2': (nan, 1.0),\n",
       " 'SUC_ACONITATE_mc1': (nan, 1.0),\n",
       " 'PHOSPHORIBULOKINASE_RXN_p1': (-0.7015678323149952, 7.691981733059328e-06),\n",
       " 'sHIS_biomass2': (nan, 1.0),\n",
       " 'sHIS_biomass1': (nan, 1.0),\n",
       " 'DIAMINOPIMEPIM_RXN_p1': (-0.6660351824254975, 3.17212220435814e-05),\n",
       " 'DIAMINOPIMEPIM_RXN_p2': (nan, 1.0),\n",
       " 'RXN0_884_p2': (nan, 1.0),\n",
       " 'RXN0_884_p1': (nan, 1.0),\n",
       " 'PHOSPHORIBULOKINASE_RXN_p2': (nan, 1.0),\n",
       " '1_PERIOD_1_PERIOD_1_PERIOD_39_RXN_m1': (nan, 1.0),\n",
       " '1_PERIOD_1_PERIOD_1_PERIOD_39_RXN_m2': (nan, 1.0),\n",
       " 'CDPKIN_RXN_m2': (nan, 1.0),\n",
       " 'CDPKIN_RXN_m1': (nan, 1.0),\n",
       " 'Ca_biomass2': (nan, 1.0),\n",
       " 'Ca_biomass1': (nan, 1.0),\n",
       " 'PYRUVATE_PROTON_mc1': (nan, 1.0),\n",
       " 'PYRUVATE_PROTON_mc2': (nan, 1.0),\n",
       " 'RXN_9516_p1': (nan, 1.0),\n",
       " 'RXN_9516_p2': (nan, 1.0),\n",
       " 'TRIOSEPISOMERIZATION_RXN_c1': (-0.8524287550852074, 5.960446240102461e-10),\n",
       " 'TRIOSEPISOMERIZATION_RXN_c2': (-0.787402613079811, 8.959696855000647e-08),\n",
       " 'CYTOCHROME_C_OXIDASE_RXN_mc1': (0.33136104663943106, 0.06393766845169929),\n",
       " 'ACETATE_COA_LIGASE_RXN_p1': (0.37173939413841023, 0.036175558711911915),\n",
       " 'ACETATE_COA_LIGASE_RXN_p2': (0.2273880506105541, 0.2107164942364535),\n",
       " 'CYTOCHROME_C_OXIDASE_RXN_mc2': (-0.6278425048573225, 0.00011969799407904704),\n",
       " 'ALANINE_GLYOXYLATE_AMINOTRANSFERASE_RXN_x1': (nan, 1.0),\n",
       " 'ALANINE_GLYOXYLATE_AMINOTRANSFERASE_RXN_x2': (nan, 1.0),\n",
       " 'LEU_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'FERULIC_ACID_pc2': (nan, 1.0),\n",
       " 'sLYS_biomass2': (nan, 1.0),\n",
       " 'sLYS_biomass1': (nan, 1.0),\n",
       " 'PREPHENATEDEHYDROG_RXN_p2': (nan, 1.0),\n",
       " 'PREPHENATEDEHYDROG_RXN_p1': (nan, 1.0),\n",
       " 'SEDOBISALDOL_RXN_p2': (nan, 1.0),\n",
       " 'sMET_biomass1': (nan, 1.0),\n",
       " 'sMET_biomass2': (nan, 1.0),\n",
       " 'RXN0_2301_m1': (nan, 1.0),\n",
       " 'Sucrose_ec1': (nan, 1.0),\n",
       " 'Sucrose_ec2': (nan, 1.0),\n",
       " 'Oleoyl_ACP_desaturase_p1': (nan, 1.0),\n",
       " 'Oleoyl_ACP_desaturase_p2': (nan, 1.0),\n",
       " 'ASPCARBTRANS_RXN_p1': (nan, 1.0),\n",
       " 'ATPPHOSPHORIBOSYLTRANS_RXN_p2': (-0.4980513481587128, 0.0037217659350883346),\n",
       " 'ATPPHOSPHORIBOSYLTRANS_RXN_p1': (-0.40170156642327426, 0.022672744720600838),\n",
       " 'ASPCARBTRANS_RXN_p2': (nan, 1.0),\n",
       " 'SUCROSE_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'SUCROSE_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'ISOCITRATE_ACONITATE_mc2': (nan, 1.0),\n",
       " 'ISOCITRATE_ACONITATE_mc1': (nan, 1.0),\n",
       " '2OXOGLUTARATEDEH_RXN_m2': (nan, 1.0),\n",
       " '2OXOGLUTARATEDEH_RXN_m1': (nan, 1.0),\n",
       " 'MYO_INOSITOL_OXYGENASE_RXN_c2': (nan, 1.0),\n",
       " 'MYO_INOSITOL_OXYGENASE_RXN_c1': (nan, 1.0),\n",
       " 'GLYCOGENSYN_RXN_p2': (nan, 1.0),\n",
       " 'SUC_CIT_mc1': (nan, 1.0),\n",
       " 'SUC_CIT_mc2': (nan, 1.0),\n",
       " 'GLYCOGENSYN_RXN_p1': (-0.6130549747461498, 0.0001911810923160899),\n",
       " 'RXN_12002_p2': (nan, 1.0),\n",
       " 'RXN_12002_p1': (nan, 1.0),\n",
       " '2_PERIOD_4_PERIOD_1_PERIOD_111_RXN_c2': (nan, 1.0),\n",
       " 'HISTIDINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'ACETYLORNTRANSAM_RXN_p1': (-0.6107801904783412, 0.00020503647991568482),\n",
       " 'ACETYLORNTRANSAM_RXN_p2': (nan, 1.0),\n",
       " 'OAA_CIT_mc2': (nan, 1.0),\n",
       " 'OAA_CIT_mc1': (nan, 1.0),\n",
       " 'Phytol_biosynthesis_p2': (nan, 1.0),\n",
       " 'Phytol_biosynthesis_p1': (nan, 1.0),\n",
       " '3_ISOPROPYLMALISOM_RXN_p2': (-0.8708228135036095, 9.201937869464706e-11),\n",
       " 'CTPSYN_RXN_p1': (nan, 1.0),\n",
       " 'CTPSYN_RXN_p2': (nan, 1.0),\n",
       " '3_ISOPROPYLMALISOM_RXN_p1': (-0.7277934164184738, 2.356434391787037e-06),\n",
       " 'AIRS_RXN_p2': (-0.6957983838629039, 9.813770045445885e-06),\n",
       " 'AIRS_RXN_p1': (-0.24691216332904295, 0.17307848239641),\n",
       " 'ISPH2_RXN_p2': (nan, 1.0),\n",
       " 'ISPH2_RXN_p1': (nan, 1.0),\n",
       " 'pLYS_biomass1': (nan, 1.0),\n",
       " 'pLYS_biomass2': (nan, 1.0),\n",
       " 'RXN_9518_p2': (nan, 1.0),\n",
       " 'RXN_9518_p1': (nan, 1.0),\n",
       " 'GLT_v_dielTransfer': (nan, 1.0),\n",
       " 'hexadecadienoate_biosynthesis_p2': (nan, 1.0),\n",
       " 'hexadecadienoate_biosynthesis_p1': (nan, 1.0),\n",
       " 'VAL_v_dielTransfer': (nan, 1.0),\n",
       " 'MALIC_NADP_RXN_c2': (-0.2987334952076483, 0.09674781041434258),\n",
       " 'MALIC_NADP_RXN_c1': (nan, 1.0),\n",
       " 'LEU_v_dielTransfer': (nan, 1.0),\n",
       " '4_PERIOD_2_PERIOD_1_PERIOD_58_RXN_p2': (-0.34000276458209167,\n",
       "  0.0569119910666727),\n",
       " '4_PERIOD_2_PERIOD_1_PERIOD_58_RXN_p1': (-0.12109729717553,\n",
       "  0.5091180470968111),\n",
       " 'UDPGLUCEPIM_RXN_c1': (nan, 1.0),\n",
       " 'UDPGLUCEPIM_RXN_c2': (nan, 1.0),\n",
       " 'SULFATE_ADENYLYLTRANS_RXN_p1': (-0.02003418287171704, 0.913335561791321),\n",
       " 'SULFATE_ADENYLYLTRANS_RXN_p2': (nan, 1.0),\n",
       " 'Phytol_degradation_p2': (nan, 1.0),\n",
       " 'Phytol_degradation_p1': (nan, 1.0),\n",
       " 'UREASE_RXN_c1': (nan, 1.0),\n",
       " 'UREASE_RXN_c2': (nan, 1.0),\n",
       " 'RXN_12541_c1': (nan, 1.0),\n",
       " 'RXN_12541_c2': (nan, 1.0),\n",
       " 'OROTATE_REDUCTASE_NADH_RXN_p2': (nan, 1.0),\n",
       " 'arachidate_biosynthesis_p1': (nan, 1.0),\n",
       " 'arachidate_biosynthesis_p2': (nan, 1.0),\n",
       " 'OROTATE_REDUCTASE_NADH_RXN_p1': (nan, 1.0),\n",
       " 'ATP_CITRATE_PRO_S_LYASE_RXN_c2': (nan, 1.0),\n",
       " 'GPH_RXN_p2': (nan, 1.0),\n",
       " 'GPH_RXN_p1': (-0.6969402322981076, 9.356001583469458e-06),\n",
       " 'ATP_CITRATE_PRO_S_LYASE_RXN_c1': (-0.7940965931826601, 5.81957651845515e-08),\n",
       " 'RXN_7984_p1': (nan, 1.0),\n",
       " 'RXN_7984_p2': (nan, 1.0),\n",
       " 'SUCROSE_PHOSPHATASE_RXN_c1': (-0.3932533451470359, 0.025970873552948755),\n",
       " 'SUCROSE_PHOSPHATASE_RXN_c2': (-0.3828402229867557, 0.030567181525926218),\n",
       " 'RXN_9623_p1': (nan, 1.0),\n",
       " 'Palmitate_biomass1': (nan, 1.0),\n",
       " 'Palmitate_biomass2': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_7_PERIOD_44_RXN_c2': (nan, 1.0),\n",
       " '2_PERIOD_7_PERIOD_7_PERIOD_44_RXN_c1': (nan, 1.0),\n",
       " 'TRP_pc1': (nan, 1.0),\n",
       " 'TRP_pc2': (nan, 1.0),\n",
       " 'MAL_PROTON_vc1': (nan, 1.0),\n",
       " 'MAL_PROTON_vc2': (nan, 1.0),\n",
       " 'HIS_v_dielTransfer': (nan, 1.0),\n",
       " 'LEU_PROTON_vc1': (nan, 1.0),\n",
       " 'LEU_PROTON_vc2': (nan, 1.0),\n",
       " 'Glycolate_xc1': (nan, 1.0),\n",
       " 'Glycolate_xc2': (nan, 1.0),\n",
       " 'ARG_v_dielTransfer': (nan, 1.0),\n",
       " 'MYO_INOSITOL_1OR_4_MONOPHOSPHATASE_RXN_c1': (nan, 1.0),\n",
       " 'MYO_INOSITOL_1OR_4_MONOPHOSPHATASE_RXN_c2': (nan, 1.0),\n",
       " 'OAA_ISOCITRATE_mc2': (nan, 1.0),\n",
       " '2_PERIOD_5_PERIOD_1_PERIOD_19_RXN_p1': (-0.4892005662316682,\n",
       "  0.004491315955747099),\n",
       " '2_PERIOD_5_PERIOD_1_PERIOD_19_RXN_p2': (nan, 1.0),\n",
       " 'VALINE_TRNA_LIGASE_RXN_c2': (0.5479217621288024, 0.00116949967256666),\n",
       " 'VALINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'Pi_tx2': (nan, 1.0),\n",
       " 'Pi_tx1': (nan, 1.0),\n",
       " 'F16ALDOLASE_RXN_c1': (-0.29464521673253014, 0.10162534805524817),\n",
       " 'F16ALDOLASE_RXN_c2': (-0.5263903521990075, 0.0019697618966521054),\n",
       " 'TRP_v_dielTransfer': (nan, 1.0),\n",
       " 'ACETOVANILLONE_HYDROXYLASE_c2': (nan, 1.0),\n",
       " 'RXN_13161_m1': (nan, 1.0),\n",
       " 'RXN_13161_m2': (nan, 1.0),\n",
       " 'MALATE_DEHYDROGENASE_NADP_RXN_p1': (0.10435184934659929, 0.5697817783975965),\n",
       " 'LYSINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'MANNPISOM_RXN_c2': (nan, 1.0),\n",
       " 'MANNPISOM_RXN_c1': (nan, 1.0),\n",
       " 'LYSINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " '3_HYDROXYPROPIONATE_DEHYDROGENASE_RXN_m1': (nan, 1.0),\n",
       " 'HISTCYCLOHYD_RXN_p2': (-0.521940233847401, 0.0021846492202787686),\n",
       " 'RXN_3422_c2': (nan, 1.0),\n",
       " 'RXN_3422_c1': (nan, 1.0),\n",
       " 'HISTCYCLOHYD_RXN_p1': (-0.27812118958269866, 0.12324722149047655),\n",
       " 'CPD_14927_xc2': (nan, 1.0),\n",
       " 'CPD_14927_xc1': (nan, 1.0),\n",
       " 'GLUTATHIONE_PEROXIDASE_RXN_p1': (nan, 1.0),\n",
       " 'PROTON_ATPase_c1': (nan, 1.0),\n",
       " 'PROTON_ATPase_c2': (nan, 1.0),\n",
       " 'FRUCTOKINASE_RXN_c2': (nan, 1.0),\n",
       " 'ACETOLACTREDUCTOISOM_RXN_p2': (nan, 1.0),\n",
       " 'ACETOLACTREDUCTOISOM_RXN_p1': (-0.8879306804961857, 1.2299856457202914e-11),\n",
       " 'FRUCTOKINASE_RXN_c1': (nan, 1.0),\n",
       " 'CYS_mc1': (nan, 1.0),\n",
       " 'CYS_mc2': (nan, 1.0),\n",
       " 'RXN_9539_p2': (nan, 1.0),\n",
       " 'RXN_9539_p1': (nan, 1.0),\n",
       " '1_PERIOD_8_PERIOD_5_PERIOD_1_RXN_p1': (-0.6516576856615199,\n",
       "  5.3430582861328994e-05),\n",
       " '1_PERIOD_8_PERIOD_5_PERIOD_1_RXN_p2': (nan, 1.0),\n",
       " 'WATER_vc2': (nan, 1.0),\n",
       " 'WATER_vc1': (nan, 1.0),\n",
       " '6PGLUCONOLACT_RXN_c2': (0.5723987431086268, 0.0006187834986103325),\n",
       " 'PRPPSYN_RXN_p2': (-0.8611499906864887, 2.5408351190062513e-10),\n",
       " 'PRPPSYN_RXN_p1': (-0.8521570672885106, 6.115511929739356e-10),\n",
       " 'MET_pc1': (nan, 1.0),\n",
       " 'MET_pc2': (nan, 1.0),\n",
       " 'RXN_1827_p2': (-0.3247070287312131, 0.06979649453737233),\n",
       " 'RXN_1827_p1': (-0.0019039406320377629, 0.9917485876080223),\n",
       " 'TYROSINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'TYROSINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'SUCCINATE_SEMIALDEHYDE_DEHYDROGENASE_RXN_m1': (nan, 1.0),\n",
       " 'SUCCINATE_SEMIALDEHYDE_DEHYDROGENASE_RXN_m2': (nan, 1.0),\n",
       " 'THREONINE_ALDOLASE_RXN_c1': (nan, 1.0),\n",
       " 'THREONINE_ALDOLASE_RXN_c2': (nan, 1.0),\n",
       " 'CYS_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'PHENYLALANINE_TRNA_LIGASE_RXN_c2': (nan, 1.0),\n",
       " 'PHENYLALANINE_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'TRP_TRNA_LIGASE_RXN_c1': (nan, 1.0),\n",
       " 'sGLY_biomass2': (nan, 1.0),\n",
       " 'RXN1G_320_p2': (nan, 1.0),\n",
       " 'RXN1G_320_p1': (nan, 1.0),\n",
       " 'CYS_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'HOMO_CYS_pc1': (nan, 1.0),\n",
       " 'HOMO_CYS_pc2': (nan, 1.0),\n",
       " 'ATPase_tx1': (nan, 1.0),\n",
       " 'ATPase_tx2': (nan, 1.0),\n",
       " 'pLEU_biomass2': (nan, 1.0),\n",
       " 'pLEU_biomass1': (nan, 1.0),\n",
       " 'ARGINASE_RXN_m2': (0.3647764648815487, 0.04009903053619148),\n",
       " 'ARGINASE_RXN_m1': (nan, 1.0),\n",
       " 'FERRIC_CHELATE_REDUCTASE_RXN_c1': (nan, 1.0),\n",
       " 'FERRIC_CHELATE_REDUCTASE_RXN_c2': (nan, 1.0),\n",
       " 'TIGLYLCOA_HYDROXY_RXN_m2': (nan, 1.0),\n",
       " 'TIGLYLCOA_HYDROXY_RXN_m1': (nan, 1.0),\n",
       " 'CYSTEINE_AMINOTRANSFERASE_RXN_m2': (nan, 1.0),\n",
       " 'ILE_PROTON_vc2': (nan, 1.0),\n",
       " 'ILE_PROTON_vc1': (nan, 1.0),\n",
       " 'CYSTEINE_AMINOTRANSFERASE_RXN_m1': (nan, 1.0),\n",
       " 'PLASTOQUINOL_PLASTOCYANIN_REDUCTASE_RXN_p1': (-0.7355718631212239,\n",
       "  1.6162804994744276e-06),\n",
       " 'PLASTOQUINOL_PLASTOCYANIN_REDUCTASE_RXN_p2': (nan, 1.0),\n",
       " 'HISTOLDEHYD_RXN_p1': (nan, 1.0),\n",
       " 'L_CITRULLINE_pc1': (nan, 1.0),\n",
       " 'RXN_7800_p2': (nan, 1.0),\n",
       " 'RXN_7800_p1': (nan, 1.0),\n",
       " '5_METHYL_THF_pc2': (nan, 1.0),\n",
       " '5_METHYL_THF_pc1': (nan, 1.0),\n",
       " 'NADPHoxm_tx2': (nan, 1.0),\n",
       " 'NITRATE_PROTON_rev_vc1': (nan, 1.0),\n",
       " 'NITRATE_PROTON_rev_vc2': (nan, 1.0),\n",
       " 'OAA_SO4_mc1': (nan, 1.0),\n",
       " 'OAA_SO4_mc2': (nan, 1.0),\n",
       " 'THRESYN_RXN_p2': (nan, 1.0),\n",
       " 'THRESYN_RXN_p1': (-0.40538938037522837, 0.021345539095741387),\n",
       " 'CA2_rev_vc1': (nan, 1.0),\n",
       " 'PROTONATP_rev_vc1': (nan, 1.0),\n",
       " 'PROTONATP_rev_vc2': (nan, 1.0),\n",
       " ...}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PCCdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "fout = open(\"Data/PPC_transcripts_fluxes_lightNotForced_constMaint_notAverage_ignoreComp_27_12_19.tsv\",\"w\")\n",
    "\n",
    "fout.write(\"Rxn ID\\treaction\\tPCC\\tp-value\\tfluxes\\n\")\n",
    "for rxn in PCCdict.keys():\n",
    "    xlist = list()\n",
    "    ylist = list()\n",
    "    for i in range(1,6):\n",
    "        xlist.append(modelDictNonCompartmentalized[i][str(rxn[:rxn.rindex(\"_\")]+rxn[len(rxn)-1]).replace(\"_NADP\",\"\").replace(\"_NAD\",\"\")])\n",
    "        ylist.append(totalTranscriptDict[i][rxn])\n",
    "    fout.write(rxn+\"\\t\"+leaf_model.reactions.get_by_id(rxn).reaction+\"\\t\"+str(PCCdict[rxn][0])+\"\\t\"+str(PCCdict[rxn][1])+\"\\t\"+str(xlist)+\"\\n\")\n",
    "fout.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "frozenset({<Gene Glyma.14G009600 at 0x7fa1d6394550>,\n",
       "           <Gene Glyma.02G304200 at 0x7fa1d6394590>,\n",
       "           <Gene Glyma.04G173400 at 0x7fa1d63945d0>,\n",
       "           <Gene Glyma.04G144300 at 0x7fa1d6394610>,\n",
       "           <Gene Glyma.19G068500 at 0x7fa1d6394650>,\n",
       "           <Gene Glyma.04G157000 at 0x7fa1d6394690>,\n",
       "           <Gene Glyma.03G190500 at 0x7fa1d63946d0>,\n",
       "           <Gene Glyma.19G190900 at 0x7fa1d6394710>,\n",
       "           <Gene Glyma.15G235300 at 0x7fa1d6394750>,\n",
       "           <Gene Glyma.06G208200 at 0x7fa1d6394790>,\n",
       "           <Gene Glyma.16G204600 at 0x7fa1d63947d0>,\n",
       "           <Gene Glyma.09G153900 at 0x7fa1d6394810>,\n",
       "           <Gene Glyma.15G207600 at 0x7fa1d6394850>})"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxn = leaf_model.reactions.get_by_id(rxn)\n",
    "rxn.genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot fluxes which show +ve correlation with transcript levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxnlist = [\"2PGADEHYDRAT_RXN_c1\"]\n",
    "\n",
    "colordict = [\"white\",\"black\",\"blue\",\"green\",\"red\",\"cyan\"]\n",
    "\n",
    "for rxn in rxnlist:\n",
    "    max_x = 0\n",
    "    min_x = 0\n",
    "    for i in range(1,6):\n",
    "        xlist = list()\n",
    "        ylist = list()\n",
    "        for j in range(0,len(totalTranscriptDict[i][rxn])):\n",
    "            xlist.append(modelDictNonCompartmentalized[i][rxn[:rxn.rindex(\"_\")]+rxn[len(rxn)-1].replace(\"NADP\",\"\").replace(\"NAD\",\"\")])\n",
    "            ylist.append(totalTranscriptDict[i][rxn][j])\n",
    "        import matplotlib.pyplot as plt\n",
    "        plt.scatter(xlist,ylist,color=colordict[i],label=\"Scenario \"+str(i))\n",
    "        if max_x<max(xlist):\n",
    "            max_x = max(xlist)\n",
    "        if min_x>min(xlist):\n",
    "            min_x = max(xlist)\n",
    "    plt.title(rxn)\n",
    "    plt.xlabel(\"predicted flux\")\n",
    "    plt.ylabel(\"transcript abundance\")\n",
    "    plt.legend()\n",
    "    if abs(min_x) > abs(max_x):\n",
    "        temp = min_x\n",
    "    else:\n",
    "        temp = max_x\n",
    "    plt.xlim(min_x-(temp*(10.0/100)),max_x+(temp*(10.0/100)))\n",
    "    plt.savefig(\"Data/Plots_27_12_2019/\"+rxn+\".svg\")\n",
    "    plt.close()\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "- Kannan, K., Wang, Y., Lang, M., Challa, G. S., Long, S. P., & Marshall-Colon, A. (2019). Combining gene network, metabolic and leaf-level models shows means to future-proof soybean photosynthesis under rising CO2. In Silico Plants, 1(1), 1–18. https://doi.org/10.1093/insilicoplants/diz008\n",
    "- Haile, F. J., & Higley, L. G. (2003). Changes in Soybean Gas-Exchange After Moisture Stress and Spider Mite Injury. Environmental Entomology, 32(3), 433–440. https://doi.org/10.1603/0046-225x-32.3.433\n",
    "- Turner, G. W., Cuthbertson, D. J., Voo, S. S., Settles, M. L., Grimes, H. D., & Lange, B. M. (2012). Experimental sink removal induces stress responses, including shifts in amino acid and phenylpropanoid metabolism, in soybean leaves. Planta, 235(5), 939–954. https://doi.org/10.1007/s00425-011-1551-4\n",
    "- Leakey, A. D. B., Xu, F., Gillespie, K. M., McGrath, J. M., Ainsworth, E. A., & Ort, D. R. (2009). Genomic basis for stimulated respiration by plants growing under elevated carbon dioxide. Proceedings of the National Academy of Sciences of the United States of America, 106(9), 3597–3602. https://doi.org/10.1073/pnas.0810955106\n",
    "- Weston, D. J., Karve, A. A., Gunter, L. E., Jawdy, S. S., Yang, X., Allen, S. M., & Wullschleger, S. D. (2011). Comparative physiology and transcriptional networks underlying the heat shock response in Populus trichocarpa, Arabidopsis thaliana and Glycine max. Plant, Cell and Environment, 34(9), 1488–1506. https://doi.org/10.1111/j.1365-3040.2011.02347.x\n",
    "- Bunce, J. (2019). Consistent Differences in Field Leaf Water-Use Efficiency among Soybean Cultivars. Plants, 8(5), 123. https://doi.org/10.3390/plants8050123"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
